---

## 🧠 🇸🇦 مقدمة القسم الثالث | المعالجة البسيطة

## 🧠 🇬🇧 Section 3 Introduction | Basic Text Processing

> 🇸🇦
> بعد أن تعرّفنا في الأقسام السابقة على الأدوات الأساسية لمعالجة اللغة (مثل Tokenization، POS، NER...)،
> ننتقل الآن إلى مرحلة **المعالجة البسيطة للنصوص (Basic Text Processing)**، وهي الخطوة التي تسبق أي تحليل ذكي أو تدريب نموذج تعلم آلي.
> هنا نبدأ بفهم **معاني الكلمات وعلاقاتها** — لأن الحاسوب لا يفهم اللغة مثل البشر، بل يحتاج إلى تمثيل منطقي يربط بين الكلمات، معانيها، ومرادفاتها.

> 🇬🇧
> After learning the core NLP tools (Tokenization, POS, NER...),
> we now move to **Basic Text Processing** — the foundation before training any intelligent model.
> This step focuses on understanding **word meanings and relationships**, since computers need structured ways to connect words, meanings, and contexts.

---

## 🔹 🇸🇦 معنى الكلمة (Word Meaning)

---

### 📘 🇸🇦 Lemma Form | 🇬🇧 The Base Form

> 🇸🇦
> "الـ Lemma" هي الشكل الأساسي أو الجذر المعجمي للكلمة بعد إزالة أي تغييرات صرفية.
> مثلًا:
>
> * book → هو الـ lemma من: booked, booking
> * run → من: running, ran
>
> 🇬🇧
> The **lemma form** is the base or dictionary form of a word — after removing tense or plural inflections.
> Example:
>
> * *book* → booked, booking
> * *run* → running, ran

---

### 📘 🇸🇦 Word Form | 🇬🇧 Inflected Word

> 🇸🇦
> "الـ Wordform" هي الكلمة بعد أن تمت إضافة التغييرات عليها (زمن، جمع، تصريف).
> مثل: *bookings*, *runner*, *sung*.
>
> 🇬🇧
> A **word form** is the inflected version of the lemma — modified by tense, plural, or derivation.
> Example: *bookings*, *runner*, *sung*.

---

### 📘 🇸🇦 تعدد المعاني | 🇬🇧 Multiple Meanings

> 🇸🇦
> الكلمة الواحدة يمكن أن تحمل أكثر من معنى حسب السياق.
> مثل كلمة **"run"** التي قد تعني "يجري"، أو "يدير"، أو "يشغّل".
> لهذا السبب، يحتاج الحاسوب إلى تحليل السياق قبل أن يقرر أي معنى هو المقصود.

> 🇬🇧
> A single word can have multiple meanings depending on the context.
> For example, **"run"** can mean "to jog", "to manage", or "to operate".
> That’s why contextual understanding is essential in NLP.

---

## 🧩 🇸🇦 المصطلحات الأساسية في المعاني | 🇬🇧 Key Word Meaning Concepts

---

### 🔹 Homographs | الكلمات المتشابهة كتابةً

> 🇸🇦
> هي الكلمات التي تُكتب بنفس الشكل لكن تحمل معاني مختلفة، مثل:
> *book*, *bank*, *run*.
> مثال:
>
> * Ahmed runs 2 km daily.
> * Ahmed runs the restaurant.
>
> هذه الجملة تُظهر كيف أن الكلمة نفسها تغيّر معناها حسب السياق.
>
> 🇬🇧
> **Homographs** are words that share the same spelling but have different meanings.
> Example:
>
> * Ahmed runs 2 km daily.
> * Ahmed runs the restaurant.
>   The meaning of *run* changes based on its context.

---

### 🔹 Homophones | الكلمات المتشابهة نطقًا

> 🇸🇦
> كلمات تُنطق بنفس الشكل ولكن تُكتب بشكل مختلف ولها معانٍ مختلفة:
>
> * *peace* / *piece*
> * *right* / *write*
>
> هذه الكلمات تسبب مشاكل أحيانًا في أنظمة **Text-to-Speech** لأنها تعتمد على الصوت فقط.
>
> 🇬🇧
> **Homophones** are words that sound the same but are spelled differently and mean different things.
> Examples:
>
> * *peace* vs. *piece*
> * *right* vs. *write*
>
> These often confuse **speech systems**, since pronunciation alone isn’t enough to disambiguate meaning.

---

### 🔹 Synonyms | المرادفات

> 🇸🇦
> الكلمات التي يمكن أن تحل محل بعضها في أغلب المواضع:
>
> * *big* ↔ *large*
> * *car* ↔ *automobile*
>
> لكن نادرًا ما تكون كلمتان متطابقتين تمامًا في المعنى؛ فـ “big brother” لا يمكن استبدالها بـ “large brother”.
>
> 🇬🇧
> **Synonyms** are words with similar meanings that can often replace each other:
>
> * *big* ↔ *large*
> * *car* ↔ *automobile*
>
> Yet, perfect synonymy is rare — for example, “big brother” ≠ “large brother”.

---

### 🔹 Antonyms | الأضداد

> 🇸🇦
> هي الكلمات التي تحمل معنى معاكسًا لكلمة أخرى، مثل:
>
> * *dark* ↔ *light*
> * *short* ↔ *long*
>
> وأحيانًا تختلف الأضداد حسب السياق، فمثلاً كلمة *big* ضدها *small* في الحجم، لكن *little* في العمر.
>
> 🇬🇧
> **Antonyms** are words with opposite meanings:
>
> * *dark* ↔ *light*
> * *short* ↔ *long*
>
> The antonym depends on context — *big* vs. *small* (size), or *big* vs. *little* (age).

---

### 🔹 Hyponyms & Hypernyms | الفروع والأصول الدلالية

> 🇸🇦
> الـ **Hyponym** هو الكلمة “الفرعية” ضمن فئة أوسع (Hypernym).
>
> * 🍓 الفراولة → Hyponym من الفواكه
> * 🚗 السيارة → Hyponym من المركبات
>
> والعكس، الفواكه والمركبات هي **Hypernyms** لتلك الكلمات.
>
> 🇬🇧
> A **Hyponym** is a more specific term within a broader category (**Hypernym**).
>
> * 🍓 *Strawberry* → Hyponym of *Fruit*
> * 🚗 *Car* → Hyponym of *Vehicle*
>
> Conversely, *Fruit* and *Vehicle* are **Hypernyms**.

---

## 🧠 🇸🇦 لماذا هذا مهم في عام 2025؟

## 🧠 🇬🇧 Why This Still Matters in 2025

> 🇸🇦
> هذه المفاهيم ليست لغوية فقط، بل أساسية في تدريب نماذج اللغة الحديثة.
> في عام 2025، يتم استخدام العلاقات بين الكلمات (Synonyms, Hypernyms...) داخل نماذج **التمثيل الدلالي** مثل
> **Word2Vec، GloVe، BERT، وAraBERT** لتقدير المعنى بناءً على السياق.
>
> أيضًا، تساعد هذه العلاقات في تحسين **تحليل المشاعر (Sentiment Analysis)**، و**المحادثات الذكية (Chatbots)**،
> و**البحث الدلالي (Semantic Search)** المستخدم في محركات مثل Google وOpenAI.

> 🇬🇧
> These semantic relationships power modern NLP models.
> In **2025**, systems like **BERT, AraBERT, and GloVe** use synonymy, antonymy, and hypernymy to model meaning in vector space.
> They’re crucial for tasks like **Sentiment Analysis**, **Chatbots**, and **Semantic Search**,
> enabling machines to infer meaning even when exact words differ.

---

## 🪄 🇸🇦 خلاصة القسم | 🇬🇧 Section Summary

> 🇸🇦
> تعلمنا في هذا القسم أن فهم معاني الكلمات هو حجر الأساس لأي معالجة لغوية.
> فقبل أن “يفهم” الحاسوب النص، يجب أن يعرف العلاقة بين الكلمات — من هي الجذر، من هي الفرع، ومن تشترك في المعنى أو تعارضه.
>
> 🇬🇧
> In summary, understanding word meaning and relationships is the foundation of NLP.
> Before a computer can “understand” text, it must know how words relate — which share meaning, which oppose, and which belong together.

---

---

## 📘 القسم الثاني : 🇸🇦 تضمين الكلمات | 🇬🇧 Word Embedding

---

### 🇸🇦 المفهوم الأساسي

تضمين الكلمات **(Word Embedding)** هو من أهم المفاهيم في معالجة اللغة الطبيعية (NLP).
الفكرة ببساطة: كل كلمة يتم تحويلها إلى **مجموعة أرقام (Vector)** تمثل معناها في فضاء عددي،
وهذه الأرقام تُعبّر عن علاقات الكلمة مع غيرها — مدى قربها أو بعدها من كلمات أخرى.

تخيل أن لدينا كلمات مثل:
**"صبر" – "رجل" – "تفاحة" – "كلب" – "كتاب"**
يمكننا طرح مجموعة أسئلة تصف هذه الكلمات مثل:

| السؤال               | صبر | رجل | تفاحة | كلب | كتاب |
| -------------------- | --- | --- | ----- | --- | ---- |
| هل هذا الشيء حي؟     | لا  | نعم | لا    | نعم | لا   |
| هل يستطيع التحدث؟    | لا  | نعم | لا    | لا  | لا   |
| هل هو ملموس؟         | لا  | نعم | نعم   | نعم | نعم  |
| هل يمكن أكله؟        | لا  | لا  | نعم   | لا  | لا   |
| هل يمكن بيعه وشراءه؟ | لا  | لا  | نعم   | نعم | نعم  |

كل "نعم" و"لا" يمكن تحويلها إلى قيم رقمية (مثل 1 و0).
وهكذا نكوّن **مصفوفة رقمية (Embedding Matrix)** تمثل كل كلمة بناءً على خصائصها.
لكن في الواقع، التضمين الحديث لا يعتمد على أسئلة محدودة كهذه،
بل على **ملايين "الأسئلة غير المباشرة"** التي يتعلمها النموذج من النصوص الضخمة.
بمعنى آخر:
كل كلمة لها "بصمة رقمية" خاصة بها — تشبه **الـDNA اللغوي** للكلمة. 🧬

كل رقم في هذا المتجه (vector) يمثل "إجابة" على سؤال ضمني عن الكلمة،
ليس فقط نعم أو لا (0 أو 1)، بل قيم نسبية مثل **0.2 أو 0.87**،
تعكس درجة التشابه أو الارتباط بمعنى معين (كالقوة، المشاعر، الزمن، إلخ).

كلما كانت هذه الأرقام أكثر دقة، أصبح المتجه أكثر قدرة على تمثيل المعنى الحقيقي للكلمة،
وبالتالي يمكننا معرفة مدى تشابهها أو اختلافها عن كلمات أخرى.

> ⚙️ مثلًا:
> يمكن للنموذج معرفة أن العلاقة بين "ملك" و"ملكة"
> تشبه العلاقة بين "رجل" و"امرأة" — بشكل أوتوماتيكي تمامًا من خلال الحسابات الرياضية داخل فضاء الكلمات.

كما أن المسافة بين الكلمات لا تُقاس فقط بالبعد (Euclidean Distance)،
لأن المسافة قد تكون خادعة أحيانًا، بل الأفضل قياس **الزاوية بين المتجهات** عبر
**Cosine Similarity** لتحديد مدى التشابه في الاتجاه والمعنى.

أما بالنسبة للجمل، فالـ *Word Embedding* لا يكون دقيقًا جدًا،
لأنه يقيس تشابه الكلمات، وليس بالضرورة **المعنى العام للسياق**.
على سبيل المثال، الجملتان:

> “I love school” و “I hate school”
> قد تكونان متشابهتين رقميًا، رغم اختلاف المعنى، لأن كلتيهما تتحدثان عن المدرسة والمشاعر.

---

### 🇬🇧 The Basic Concept

**Word Embedding** is one of the fundamental ideas in NLP.
It transforms each word into a **vector of numbers** that captures its meaning and relationships with other words.
These vectors act like the **word’s DNA**, encoding its semantic traits numerically. 🧬

You can imagine this as answering a massive set of “hidden questions” —
not just simple *yes/no* ones, but nuanced, continuous values like 0.27 or 0.89
that describe various aspects of meaning such as tone, context, or sentiment.

Thus, every word gets a **unique numerical fingerprint**,
allowing models to compute how close or far words are from each other in meaning.

> ⚙️ For example, the model can figure out relationships like:
> **king – man + woman ≈ queen**
> automatically, because the embeddings capture deep semantic relationships.

However, word embeddings aren’t perfect for **sentence-level similarity**.
Two opposite sentences — like *“I love school”* and *“I hate school”* —
might still appear similar numerically, because both discuss the same topic (*school*)
and share emotional or contextual proximity.

When measuring similarity, using **Cosine Similarity** is preferred over Euclidean distance,
since cosine focuses on the *direction* (meaning) rather than just numerical distance,
providing more accurate results in semantic space.

---

---

## 📊 🇸🇦 الفرق بين Euclidean Distance و Cosine Similarity

<p align="center">
  <img src="assets/cosine & Euclidean.png" alt="NLP Tools" width="400">
</p>

في الصورة عندك ثلاث مقالات (أو كوربس):

* **Food corpus** عند النقطة (5, 15)
* **Agriculture corpus** عند النقطة (20, 40)
* **History corpus** عند النقطة (30, 20)

كل واحدة مُمثَّلة كـ **متجه (vector)** في فضاء ثنائي الأبعاد:
المحور الأفقي = كلمة *disease*
المحور العمودي = كلمة *eggs*

### 1️⃣ ماذا يفعل الـ Euclidean Distance؟

**Euclidean distance** هو المسافة المستقيمة بين نقطتين (زي المسطرة بين نقطتين على الورقة).

في الصورة:

* (d_1) = المسافة بين **Food** و **Agriculture**
* (d_2) = المسافة بين **Agriculture** و **History**

النتيجة في الرسم:

> (d_2 < d_1)
> أي أن **Euclidean distance** يقول إن مقالة **History** أقرب إلى **Agriculture** من مقالة **Food**.

لكن **منطقيًا**؟ 🤔
مقالة عن **Food** (طعام) غالبًا أقرب لمقالة عن **Agriculture** (زراعة)
من مقالة عن **History** (تاريخ).

إذن هنا **المسافة وحدها خدّاعة**؛ لأنها تأثرت بطول المتجه (حجم الأرقام)،
وليس *فكرة المتجه واتجاهه*.

---

### 2️⃣ ماذا يفعل الـ Cosine Similarity؟

**Cosine similarity** لا يهتم بطول المتجه، بل **بالزاوية بين المتجهين**:

* الزاوية **α** بين متجه **Food** و **Agriculture** صغيرة
* الزاوية **β** بين متجه **History** و **Agriculture** أكبر (β > α)

كلما كانت الزاوية أصغر → الاتجاهان متشابهان أكثر → الموضوعان أقرب معنويًا.
لذلك **Cosine similarity** سيعتبر:

> Food أقرب إلى Agriculture من History

وهذا يطابق **المنطق والمعنى** 👍

---

### 3️⃣ قيم Cosine ماذا تعني بالضبط؟

Cosine similarity يحسب:

[
\cos(\theta) = \frac{\text{dot product}}{|\vec{a}| \cdot |\vec{b}|}
]

حيث θ هي الزاوية بين المتجهين.
المهم هي **قيمة cos(θ)**:

| الزاوية θ بين المتجهين | قيمة cosine | التفسير                                 |
| ---------------------- | ----------- | --------------------------------------- |
| 0°                     | 1           | متطابقان في الاتجاه → تشابه تام تقريبًا |
| 30° تقريبًا            | ~0.87       | تشابه قوي جدًا                          |
| 60° تقريبًا            | ~0.5        | تشابه متوسط                             |
| 90°                    | 0           | لا يوجد تشابه (متعامدان)                |
| > 90° حتى 180°         | < 0         | علاقة عكسية / اتجاه معاكس               |

فلو قلنا:

* **cosine = 0.9** → النصان/الكلمتان شبه متطابقتين معنويًا
* **cosine = 0.4** → في شبه بس مو قوي
* **cosine = 0** → تقريبًا ما بينهم علاقة
* **cosine = -0.5** → يميلان لاتجاه معاكس (أقرب لضد/تناقض في بعض الأبعاد)

لهذا السبب في الـ **Word Embeddings** غالبًا نستخدم **Cosine similarity** بدل Euclidean distance،
خصوصًا في النصوص، لأن طول المتجه يتأثر بأشياء مثل **طول الجملة أو تكرار الكلمات**،
بينما الزاوية تعكس **المعنى والاتجاه** أكثر من الحجم.

---

## 📊 🇬🇧 Euclidean Distance vs. Cosine Similarity

<p align="center">
  <img src="assets/cosine & Euclidean.png" alt="NLP Tools" width="400">
</p>

In the figure, we have three corpora represented as vectors:

* **Food corpus** at (5, 15)
* **Agriculture corpus** at (20, 40)
* **History corpus** at (30, 20)

Each point is a **vector** in a 2D space:
x-axis = frequency of *“disease”*
y-axis = frequency of *“eggs”*.

### 1️⃣ What does Euclidean Distance do?

**Euclidean distance** is just the straight-line distance between two points.

In the plot:

* (d_1) = distance between **Food** and **Agriculture**
* (d_2) = distance between **Agriculture** and **History**

The picture shows:

> (d_2 < d_1)

So Euclidean distance claims **History** is closer to **Agriculture** than **Food** is.

But semantically? 🤔
A **Food** article is usually more related to **Agriculture** than a **History** article.
So distance alone can be **misleading**, because it’s affected by **vector length**,
not only the *direction* of meaning.

---

### 2️⃣ What does Cosine Similarity do?

**Cosine similarity** ignores the length and looks at the **angle between vectors**:

* Angle **α** between **Food** and **Agriculture** is small
* Angle **β** between **History** and **Agriculture** is larger (β > α)

Smaller angle → more similar direction → more similar topic.
So cosine similarity will say:

> Food is more similar to Agriculture than History is

which matches our **intuitive understanding** 👍

---

### 3️⃣ What do cosine values actually mean?

Cosine similarity is defined as:

[
\cos(\theta) = \frac{\vec{a} \cdot \vec{b}}{|\vec{a}| \cdot |\vec{b}|}
]

Where θ is the angle between the two vectors.
What matters is the **cosine value**:

| Angle θ between vectors | Cosine value | Interpretation                            |
| ----------------------- | ------------ | ----------------------------------------- |
| 0°                      | 1            | Same direction → almost identical meaning |
| ~30°                    | ~0.87        | Very strong similarity                    |
| ~60°                    | ~0.5         | Moderate similarity                       |
| 90°                     | 0            | No similarity (orthogonal)                |
| > 90° to 180°           | < 0          | Opposite / contradictory directions       |

So:

* **cosine = 0.9** → very similar
* **cosine = 0.4** → somewhat related
* **cosine = 0** → unrelated
* **cosine = -0.5** → somewhat opposite in meaning

That’s why in **word embeddings** and text similarity,
we usually prefer **cosine similarity** over Euclidean distance:
length can be influenced by sentence length or word frequency,
but the **angle** better reflects semantic direction and meaning.

---

In [ ]:
# !pip uninstall -y spacy thinc blis numpy
# !pip install numpy==1.26.4 blis==0.7.11 thinc==8.2.2 spacy==3.7.2
# !python -m spacy download en_core_web_lg

Found existing installation: spacy 3.7.2
Uninstalling spacy-3.7.2:
  Successfully uninstalled spacy-3.7.2
Found existing installation: thinc 8.2.2
Uninstalling thinc-8.2.2:
  Successfully uninstalled thinc-8.2.2
Found existing installation: blis 0.7.11
Uninstalling blis-0.7.11:
  Successfully uninstalled blis-0.7.11
Found existing installation: numpy 1.26.4
Uninstalling numpy-1.26.4:
  Successfully uninstalled numpy-1.26.4
  Using cached numpy-1.26.4-cp310-cp310-win_amd64.whl.metadata (61 kB)
  Using cached blis-0.7.11-cp310-cp310-win_amd64.whl.metadata (7.6 kB)
  Using cached thinc-8.2.2-cp310-cp310-win_amd64.whl.metadata (15 kB)
  Using cached spacy-3.7.2-cp310-cp310-win_amd64.whl.metadata (26 kB)
Using cached numpy-1.26.4-cp310-cp310-win_amd64.whl (15.8 MB)
Using cached blis-0.7.11-cp310-cp310-win_amd64.whl (6.6 MB)
Using cached thinc-8.2.2-cp310-cp310-win_amd64.whl (1.5 MB)
Using cached spacy-3.7.2-cp310-cp310-win_amd64.whl (12.1 MB)

   ---------------------------------------- 0/4

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
ultralytics 8.3.40 requires matplotlib>=3.3.0, which is not installed.
ultralytics 8.3.40 requires pandas>=1.1.4, which is not installed.
ultralytics 8.3.40 requires pillow>=7.1.2, which is not installed.
ultralytics 8.3.40 requires pyyaml>=5.3.1, which is not installed.
ultralytics 8.3.40 requires scipy>=1.4.1, which is not installed.
ultralytics 8.3.40 requires seaborn>=0.11.0, which is not installed.
ultralytics 8.3.40 requires torch!=2.4.0,>=1.8.0; sys_platform == "win32", which is not installed.
ultralytics 8.3.40 requires torch>=1.8.0, which is not installed.
ultralytics 8.3.40 requires torchvision>=0.9.0, which is not installed.
ultralytics-thop 2.0.12 requires torch, which is not installed.
scikit-learn 1.7.2 requires scipy>=1.8.0, which is not installed.
scikit-learn 1.7.2 requires threadpoolctl>=3.1.0, w

  Using cached https://github.com/explosion/spacy-models/releases/download/en_core_web_lg-3.7.1/en_core_web_lg-3.7.1-py3-none-any.whl (587.7 MB)
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_lg')


In [1]:
# Download and load the large English model
# !python -m spacy download en_core_web_lg

import spacy
nlp = spacy.load('en_core_web_lg')


In [2]:
# Display the 300-dimensional vector for a given word
nlp(u'brave').vector


array([ 1.2657e+00,  3.0703e+00, -1.7375e+00, -2.0443e+00, -1.1928e+00,
        4.7458e+00,  4.1327e-03, -9.9579e-01,  2.7430e+00,  1.5209e+00,
       -1.5743e+00, -1.0238e+00, -4.1670e-01,  3.0210e+00,  3.3419e+00,
       -1.4513e+00,  1.3891e+00, -2.6773e+00,  1.5304e+00, -1.2868e+00,
        2.2221e+00, -1.8497e-01, -3.6390e-01, -2.1370e+00,  3.3337e+00,
        4.2660e-01,  2.4932e+00, -1.4288e+00,  2.7081e+00,  3.7910e-01,
        3.6824e+00, -3.4750e+00,  2.6784e+00, -3.1675e+00,  7.1397e-01,
        3.7714e-01, -2.3996e+00,  3.0475e-01,  9.8345e-01, -1.6483e+00,
       -2.1484e+00,  2.0593e+00, -1.2598e+00,  5.4024e-01, -1.1401e+00,
       -2.3316e+00, -1.2535e+00, -4.0506e+00, -2.4282e-02,  3.3028e+00,
       -7.1632e-01,  4.8101e-02,  1.2775e+00, -2.2964e-02,  1.7544e+00,
       -2.6502e+00,  9.9385e-01,  3.8857e-02, -1.0332e+00, -7.2275e-01,
       -1.9981e+00, -2.3808e+00, -9.7850e-02, -7.5982e-01, -6.7057e-01,
        5.1752e-01,  2.4360e-01, -3.1495e+00,  2.9339e+00, -9.42

In [3]:
# Check the vector length (should be 300)
len(nlp(u'lion').vector)


300

In [4]:
# Sentence vector (average of all word vectors)
doc = nlp(u'The quick brown fox jumped over the lazy dogs.')
doc.vector


array([-1.7769655 ,  0.39714497, -1.695121  , -0.1089559 ,  3.861494  ,
       -0.10778303, -0.02750097,  3.191314  ,  1.0857747 , -0.2615487 ,
        4.0720797 ,  1.5932049 , -2.7569218 ,  0.70982707,  2.0976841 ,
        0.08150103,  0.8847861 , -0.505237  ,  0.767067  , -2.88911   ,
       -0.28514975, -0.331664  ,  0.306348  , -2.25347   ,  0.96798134,
       -0.030282  , -3.765162  , -2.168157  ,  1.3985709 ,  2.175709  ,
       -0.81103534, -0.55156004, -1.033463  , -2.3130198 , -2.892054  ,
       -2.843568  , -0.33247897,  1.620013  ,  3.03307   , -0.42730814,
        1.298548  ,  0.18969259,  1.234282  , -0.14263602, -1.427765  ,
       -0.05807757,  0.33836406, -1.6987331 , -2.13661   ,  0.10412004,
        0.62479395,  3.9712129 , -0.31110606, -1.9676571 , -0.11860895,
        0.55582994, -0.660888  ,  1.947435  ,  1.6391805 ,  0.6569032 ,
        0.054408  , -2.08993   ,  1.0370519 ,  0.5363236 ,  0.00807395,
       -0.91060096, -3.3870788 , -1.4823462 ,  1.4170542 ,  0.32

In [5]:
# Compare similarity between multiple words
tokens = nlp(u'lion cat pet')

for token1 in tokens:
    for token2 in tokens:
        print(token1.text, token2.text, token1.similarity(token2))


lion lion 1.0
lion cat 0.3854507803916931
lion pet 0.20031583309173584
cat lion 0.3854507803916931
cat cat 1.0
cat pet 0.732966423034668
pet lion 0.20031583309173584
pet cat 0.732966423034668
pet pet 1.0


In [6]:
# Word-to-word similarity examples
print(nlp(u'man').similarity(nlp(u'woman')))
print(nlp(u'man').similarity(nlp(u'tree')))
print(nlp(u'lion').similarity(nlp(u'tiger')))
print(nlp(u'lion').similarity(nlp(u'dandelion')))


0.8273442557707831
0.24762304993106887
0.7136095195259959
0.4145525929565585


In [7]:
# Sentence similarity comparison
print(nlp(u'I love school').similarity(nlp(u'I hate school')))
print(nlp(u'this film is awesome. I love it').similarity(nlp(u'this film is boring. I hate it')))


0.9503183859344599
0.9793199767254251


In [8]:
# Check vector existence, vector norm, and OOV (out of vocabulary)
tokens = nlp(u'dog cat nargle hesham lksdvsk')

for token in tokens:
    print(token.text, token.has_vector, token.vector_norm, token.is_oov)


dog True 75.254234 False
cat True 63.188496 False
nargle False 0.0 True
hesham False 0.0 True
lksdvsk False 0.0 True


In [ ]:
# !pip install --ignore-installed --no-deps numpy==1.25.2
# !pip install --ignore-installed --no-deps scipy==1.15.3


  Using cached numpy-1.25.2-cp310-cp310-win_amd64.whl.metadata (5.7 kB)
Using cached numpy-1.25.2-cp310-cp310-win_amd64.whl (15.6 MB)
  Using cached scipy-1.15.3-cp310-cp310-win_amd64.whl.metadata (60 kB)
Using cached scipy-1.15.3-cp310-cp310-win_amd64.whl (41.3 MB)


ERROR: Could not install packages due to an OSError: [WinError 5] Access is denied: 'c:\\Users\\shosh\\.conda\\envs\\NLP\\Lib\\site-packages\\scipy\\spatial\\_distance_wrap.cp310-win_amd64.pyd'
Consider using the `--user` option or check the permissions.



In [9]:
# Example: king - man + woman ≈ queen
from scipy.spatial import distance

# Define cosine similarity function
cosine_similarity = lambda x, y: 1 - distance.cosine(x, y)

king = nlp.vocab['king'].vector
man = nlp.vocab['man'].vector
woman = nlp.vocab['woman'].vector

# Compute new vector
new_vector = king - man + woman

computed_similarities = []
words = ['cat','apple','queen','castle','sea','shell','orange','phone',
         'tiffany','angry','book','white','land','study','crown','prince',
         'dog','great','princess','elizabeth','wow','eat','dead','horrible']

# Compare the new vector with all words
for word in words:
    similarity = cosine_similarity(new_vector, nlp.vocab[word].vector)
    computed_similarities.append((word, similarity))

# Sort and show the most similar words
computed_similarities = sorted(computed_similarities, key=lambda item: -item[1])
for a, b in computed_similarities[:10]:
    print(f'Word {a} , has similarity {b}')

# import numpy as np

# # Define cosine similarity using numpy
# def cosine_similarity(x, y):
#     return np.dot(x, y) / (np.linalg.norm(x) * np.linalg.norm(y))

# king = nlp.vocab['king'].vector
# man = nlp.vocab['man'].vector
# woman = nlp.vocab['woman'].vector

# # Compute new vector
# new_vector = king - man + woman

# words = ['cat','apple','queen','castle','sea','shell','orange','phone',
#          'tiffany','angry','book','white','land','study','crown','prince',
#          'dog','great','princess','elizabeth','wow','eat','dead','horrible']

# computed_similarities = []

# for word in words:
#     similarity = cosine_similarity(new_vector, nlp.vocab[word].vector)
#     computed_similarities.append((word, similarity))

# # Sort and show the most similar words
# computed_similarities = sorted(computed_similarities, key=lambda item: -item[1])
# for a, b in computed_similarities[:10]:
#     print(f'Word {a} , has similarity {b}')


Word prince , has similarity 0.6562829656127154
Word princess , has similarity 0.6480762222378031
Word queen , has similarity 0.6178013742941038
Word crown , has similarity 0.5205165367683633
Word castle , has similarity 0.3856709507986278
Word land , has similarity 0.37457909091965713
Word elizabeth , has similarity 0.3294205394308072
Word great , has similarity 0.2888008262644557
Word white , has similarity 0.2524483583823236
Word study , has similarity 0.23255463266352316


In [10]:
# Arabic embeddings are not supported in this English model
nlp(u'أسد').vector

doc = nlp(u'ذهب محمد الي الجامعة')
doc.vector


array([ 0.0832475 , -0.054465  ,  0.307     , -0.417525  , -0.11432   ,
        0.25995   , -0.1459275 ,  0.1574625 ,  0.0258875 ,  0.1498425 ,
        0.099775  ,  0.06666   ,  0.303325  , -0.04788   ,  0.229525  ,
        0.00786325, -0.07418   ,  0.15416   , -0.004022  ,  0.29705   ,
       -0.400225  ,  0.007079  ,  0.15236   ,  0.30375   , -0.038155  ,
       -0.1529925 ,  0.0514875 ,  0.3278    , -0.00083627, -0.1376675 ,
       -0.244045  ,  0.222105  ,  0.180235  , -0.0543225 ,  0.301275  ,
       -0.14083   , -0.00201715,  0.55655   , -0.218385  , -0.1699    ,
        0.223975  ,  0.1761425 ,  0.2013475 ,  0.1498725 ,  0.155475  ,
       -0.0435075 , -0.364775  ,  0.161775  , -0.0255625 ,  0.188165  ,
       -0.2082325 , -0.1933925 , -0.254425  , -0.143445  , -0.274175  ,
       -0.0765625 , -0.083485  , -0.1528975 ,  0.295275  ,  0.0310625 ,
        0.0740225 , -0.0315975 , -0.297325  ,  0.12227   ,  0.1183175 ,
       -0.088645  ,  0.0793925 ,  0.1314175 , -0.25975   ,  0.14


> 🇸🇦
> 📍القيم ستكون كلها **صفرًا**، لأن مكتبة **spaCy** لا توفر حاليًا تمثيلات *Word Embeddings* للغة العربية.
> لاحقًا سنتعرف على أدوات قوية مثل **GloVe** و **Word2Vec** و **FastText** و **AraVec**،
> والتي تقدم تمثيلات رقمية دقيقة جدًا للنصوص العربية، تساعدنا على فهم المعاني بشكل أعمق.

> 🇬🇧
> 📍All values will appear as **zeros**, because **spaCy** currently does not provide *Word Embeddings* for Arabic.
> Later, we’ll explore powerful tools like **GloVe**, **Word2Vec**, **FastText**, and **AraVec**,
> which offer highly accurate numeric representations for Arabic text, enabling deeper semantic understanding.

---




## ✨ الخاتمة

### 🇸🇦

الآن أصبحت تعرف أن الكلمات ليست مجرد حروف، بل **أرقام تعبّر عن معانٍ عميقة** يفهمها الحاسوب!
في القسم التالي، سنبدأ رحلة **المعالجة البسيطة للنصوص (Basic Text Processing)**،
لنتعلم كيف نُحضّر البيانات قبل بناء النماذج الذكية.
استعد 🔥 — المرحلة القادمة هي أول خطوة نحو فهم اللغة بعمق!

---

### 🇬🇧

Now you know that words are **not just letters**, but **vectors of meaning** that computers can understand!
In the next section, we’ll move on to **Basic Text Processing**,
where we’ll prepare our data for building smart NLP models.
Get ready 🚀 — the exciting part starts now!

---

---

## 📘 القسم الثالث: 🇸🇦 كيس الكلمات | 🇬🇧 Bag of Words (BoW)

---

### 🇸🇦 المفهوم الأساسي

تخيل أن عندك قصة قصيرة أو تغريدة، وتبغى تعطيها للكمبيوتر يفهمها... لكن الكمبيوتر ما يعرف إلا الأرقام.
فكيف نخليه يفهم الكلمات؟ 🤔
هنا يأتي دور **Bag of Words (كيس الكلمات)** — فكرة بسيطة جدًا لكنها كانت الأساس لكل نماذج اللغة القديمة.

الفكرة أنك **تحوّل النص إلى أرقام بناءً على وجود الكلمات وتكرارها** فقط،
بدون النظر للمعنى أو الترتيب.

👝 تخيّل عندك “كيس” فيه كل كلمات النصوص، وكل جملة مجرد **مجموعة كلمات موجودة أو غير موجودة** في هذا الكيس.

---

### 🇬🇧 The Basic Idea

Imagine you give your computer a sentence, but it can’t read — it only understands numbers.
How can we represent text as numbers?
That’s what **Bag of Words (BoW)** does!

It simply **counts how many times each word appears** in a text — ignoring grammar, order, or meaning.
Think of it like a big **bag of words**, where every document is represented by word presence and frequency.

---

### 🧱 🇸🇦 الخطوتان الأساسيتان | 🇬🇧 The Two Main Steps

| 🇸🇦 الخطوة                | 🇬🇧 Step                          | 🇸🇦 الشرح                                                                                           | 🇬🇧 Explanation                                                                                           |
| -------------------------- | ---------------------------------- | ---------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------- |
| 1️⃣ إنشاء قائمة بالكلمات   | Create the vocabulary              | نجمع كل الكلمات التي ظهرت في النصوص بعد تنظيفها (تحويل لحروف صغيرة – حذف التكرار – إزالة stopwords). | Collect all unique words from the dataset after cleaning (lowercasing, removing duplicates and stopwords). |
| 2️⃣ تمثيل كل جملة بالأرقام | Represent each sentence as numbers | نحدد هل الكلمة موجودة (1) أو لا (0) أو عدد مرات ظهورها.                                              | For each sentence, mark presence (1) or absence (0) of each word — or use counts if needed.                |

---

### ✏️ 🇸🇦 مثال من الأدب الكلاسيكي | 🇬🇧 Classic Example

من رواية تشارلز ديكنز **“قصة مدينتين”**

> It was the best of times, it was the worst of times, it was the age of wisdom, it was the age of foolishness...

بعد تنظيف النص من التكرار والرموز، نحصل على قائمة كلمات مثل:
**[it, was, the, best, of, times, worst, age, wisdom, foolishness]**

ثم نعمل جدول (أو مصفوفة) يوضح وجود الكلمات في كل جملة:

| الجملة                        | it | was | the | best | of | times | worst | age | wisdom | foolishness |
| ----------------------------- | -- | --- | --- | ---- | -- | ----- | ----- | --- | ------ | ----------- |
| It was the best of times      | 1  | 1   | 1   | 1    | 1  | 1     | 0     | 0   | 0      | 0           |
| It was the worst of times     | 1  | 1   | 1   | 0    | 1  | 1     | 1     | 0   | 0      | 0           |
| It was the age of wisdom      | 1  | 1   | 1   | 0    | 1  | 0     | 0     | 1   | 1      | 0           |
| It was the age of foolishness | 1  | 1   | 1   | 0    | 1  | 0     | 0     | 1   | 0      | 1           |

---



### 💬 🇸🇦 توضيح بسيط بمثال واقعي | 🇬🇧 Simple Real-World Example

🇸🇦 **تخيل عندك محادثتين:**

1. “اليوم الجو جميل جدًا والشمس مشرقة”
2. “اليوم ممطر وغائم”

كيس الكلمات (**Bag of Words**) سيعامل كل جملة كـ **قائمة من الكلمات** فقط:

> ["اليوم", "الجو", "جميل", "ممطر", "غائم", "الشمس", "مشرقة"]

كل جملة تتحول إلى **متجه (Vector)** من الأصفار والواحدات حسب وجود الكلمة.
لكن لاحظ 👀: النموذج لا يعرف أن “جميل” و“مشرق” متقاربتان في المعنى 🌤️
وهنا تظهر **عيوب BoW** لاحقًا.

---

🇬🇧 **Imagine you have two sentences:**

1. “The weather is very nice and sunny today.”
2. “Today is rainy and cloudy.”

The **Bag of Words (BoW)** model treats each sentence as just a **list of words**:

> ["today", "weather", "nice", "rainy", "cloudy", "sunny"]

Each sentence becomes a **vector of 0s and 1s** depending on which words appear.
But notice 👀: the model doesn’t understand that *“nice”* and *“sunny”* are semantically related 🌤️ —
and that’s where **BoW’s main limitations** show up.

---



### 🧠 🇸🇦 المعالجة المسبقة للنصوص | 🇬🇧 Text Preprocessing

قبل تطبيق BoW، نعمل الخطوات التالية لتحسين النتيجة:

| 🇸🇦 الخطوة            | 🇬🇧 Step          | 🇸🇦 الهدف                           | 🇬🇧 Purpose                                      |
| ---------------------- | ------------------ | ------------------------------------ | ------------------------------------------------- |
| تحويل الحروف إلى صغيرة | Lowercasing        | لتوحيد شكل الكلمات                   | Normalize word form                               |
| حذف كلمات التوقف       | Remove stopwords   | لأنها لا تضيف معنى مثل “من، في، على” | Remove uninformative words like “the”, “of”, “in” |
| حذف العلامات           | Remove punctuation | لأن الرموز لا تعني شيئًا في التحليل  | Clean non-word characters                         |

---

-

### 🧮 🇸🇦 مفهوم الـ N-Gram وعلاقته بـ Bag of Words

### 🇬🇧 The Concept of N-Gram and Its Relation to Bag of Words

---

#### 🇸🇦 أولًا: ما هي الـ N-Gram؟

الـ **N-Gram** هي ببساطة مجموعة من **الكلمات المتتالية** تُعامل كوحدة واحدة.

* عندما نأخذ **كلمة واحدة فقط** في التحليل، فهذا يسمى **Unigram**.
* عندما نأخذ **كل كلمتين متجاورتين**، فهذا يسمى **Bigram**.
* وعندما نأخذ **كل ثلاث كلمات متجاورة**، يصبح **Trigram** … وهكذا.

👀 مثال بسيط:

> الجملة: “I love natural language processing”

* Unigrams → [I], [love], [natural], [language], [processing]
* Bigrams → [I love], [love natural], [natural language], [language processing]
* Trigrams → [I love natural], [love natural language], [natural language processing]

إذًا، بدل أن ننظر إلى الكلمة لوحدها، نبدأ ننظر إلى **العلاقة بين الكلمات المتجاورة**.

---

#### 🇬🇧 What is an N-Gram?

An **N-Gram** is simply a group of **N consecutive words** treated as a single unit.

* 1 word → **Unigram**
* 2 words → **Bigram**
* 3 words → **Trigram**, and so on.

💡 Example:

> Sentence: “I love natural language processing”

* Unigrams → [I], [love], [natural], [language], [processing]
* Bigrams → [I love], [love natural], [natural language], [language processing]
* Trigrams → [I love natural], [love natural language], [natural language processing]

So instead of analyzing words individually, we analyze **word sequences** to capture short phrases and context.

---

#### 🧠 🇸🇦 علاقتها بـ Bag of Words

الـ **N-Gram** هي في الأساس **تطوير لفكرة الـ Bag of Words**.
في BoW، نحسب تكرار الكلمات منفصلة،
لكن مع N-Gram نبدأ نحسب تكرار **التسلسلات** أيضًا.
يعني بدل ما نحسب وجود كلمة “artificial” فقط،
نحسب وجود **“artificial intelligence”** ككيان واحد له معنى أقوى.

✅ **الفائدة:**

* تساعد النموذج على فهم **السياق القريب بين الكلمات**.
* مثلًا “not good” = سلبي، بينما “good” = إيجابي، والفرق يظهر فقط عند استخدام Bigram.

⚠️ **العيب:**

* كلما زاد N، يزيد عدد التركيبات الممكنة بشكل ضخم جدًا
  (يسمى هذا بـ **Dimensionality Explosion**)
  مما يجعل النموذج أبطأ وأكبر في الذاكرة.

---

#### 🇬🇧 Relation to Bag of Words

**N-Gram** is an **extension of the Bag of Words model**.
In BoW, we count single words.
In N-Gram, we count **word sequences** — capturing short context.

✅ **Benefits:**

* Helps the model understand local context (e.g., “not good” ≠ “good”).
* Improves accuracy in text classification and sentiment analysis.

⚠️ **Drawbacks:**

* As N increases, the number of combinations grows dramatically
  (**Dimensionality Explosion**).
* This increases both **memory usage** and **processing time**.

---



## 🇸🇦 هل الـ N-Gram ما زال يُستخدم؟ وهل فعلاً حل مشكلة الـ BoW؟

### 💬 الجواب المختصر:

> نعم، الـ **N-Gram** ما زال يُستخدم،
> لكنه اليوم يُعتبر **طريقة كلاسيكية** تُستخدم في بعض المهام فقط،
> بينما **النماذج الحديثة مثل Word2Vec وBERT** تجاوزته بمراحل كبيرة.

---

### 🧠 الفهم التفصيلي:

#### 🔹 أولاً: هل حل مشكلة الـ BoW؟

إلى حدٍّ ما ✅

الـ **N-Gram** جاء كحلّ مؤقت لأكبر عيب في BoW:

> أن BoW لا يفهم الترتيب أو السياق.

فبدل ما يعدّ الكلمات بشكل منفصل،
صار يقرأ **تسلسلات قصيرة من الكلمات (n كلمات)**
وبالتالي صار يقدر يلتقط شوية معنى من السياق القريب.

مثلاً:

* BoW يشوف: “not” و“good” ككلمتين منفصلتين → ويظن الجملة إيجابية 😅
* N-Gram (Bigram) يشوف “not good” كوحدة واحدة → فيفهم إنها سلبية 👎

لكن رغم هذا التطوير، **ما زال N-Gram محدود**:

* ما زال ما يفهم المعنى الحقيقي للكلمات.
* وما يقدر يميز بين “good job” و“good idea” من حيث المعنى العام.
* وكل ما زاد N، صار النموذج ثقيل جدًا (عشرات آلاف الأعمدة في المصفوفة).

---

#### 🔹 ثانيًا: هل ما زال يُستخدم؟

نعم، لكن في حالات محددة 🎯

يُستخدم اليوم في:

* النماذج الإحصائية البسيطة (مثل Markov Models و n-gram language models).
* محركات الإكمال التلقائي البسيطة (Autocomplete, Predictive Text).
* تحليل النصوص القديمة أو المشاريع التعليمية.
* وبعض تطبيقات **الـ NLP منخفضة الموارد** (Low-Resource NLP) حيث ما في بيانات ضخمة.

---

#### 🔹 ثالثًا: ماذا حلّ مكانه؟

منذ عام **2013** مع ظهور **Word2Vec** ثم **GloVe** وبعدها **BERT**،
العالم تقريبًا **انتقل إلى مرحلة “الفهم الدلالي الحقيقي”** (Semantic Understanding).

> النماذج الجديدة ما تعتمد على عدّ الكلمات،
> بل **تتعلم المعنى والسياق الكامل لكل كلمة** من خلال ملايين الجمل.

بالتالي، نقدر نقول باختصار:

> BoW → مرحلة أولى
> N-Gram → ترقيع مؤقت رائع 🧩
> Word2Vec → البداية الحقيقية للفهم
> BERT → عصر الفهم العميق للسياق الكامل للجمل 🌍

---

## 🇬🇧 In Summary (English Version)

**Is N-Gram still used?**

> Yes, but mainly for simple or classical NLP tasks.

**Did it solve BoW’s problems?**

> Partially.
> It helped capture short context (“not good”) but still failed to understand meaning deeply.

**Why it’s less common today?**

> Because modern models (Word2Vec, GloVe, FastText, BERT, GPT…)
> learn contextual relationships directly — without needing fixed N-Grams.

**In short:**

> BoW → counts words
> N-Gram → counts short word sequences
> Word2Vec → understands meaning
> BERT → understands full context

---

### 💡 🇸🇦 العيوب | 🇬🇧 Limitations

| 🇸🇦 العيب                             | 🇬🇧 Limitation                        |
| -------------------------------------- | -------------------------------------- |
| لا يفهم المعنى أو الترتيب              | Loses meaning and order of words       |
| الحجم الكبير للمصفوفة                  | Creates very large, sparse matrices    |
| حساس للتكرار والرموز                   | Sensitive to text noise and redundancy |
| لا يميّز بين الكلمات المتشابهة معنويًا | Cannot understand synonyms or context  |

---


### 🪄 🇸🇦 خلاصة بسيطة | 🇬🇧 Quick Summary

🇸🇦 **الخلاصة بالعربية:**

* 🧩 BoW هو **أول خطوة لفهم النصوص بالأرقام**.
* 📊 لكنه يعتمد فقط على **التكرار والإحصاء** وليس على **المعنى أو السياق**.
* 🚀 النماذج الحديثة مثل **Word2Vec** و **BERT** جاءت لتكمّل هذا النقص وتفهم اللغة بعمق أكبر.

---

🇬🇧 **English Summary:**

* 🧩 **BoW** was the **first step toward representing text as numbers.**
* 📊 However, it relies only on **frequency and statistics**, not **meaning or context.**
* 🚀 Modern models like **Word2Vec** and **BERT** were developed to overcome these limits and understand language more deeply.

---

### 🎯 🇸🇦 خاتمة تحفيزية للقسم القادم

> 🇸🇦
> والآن بعدما عرفنا كيف نحول النصوص إلى أرقام بسيطة باستخدام BoW،
> حان الوقت نخطو خطوة للأمام ونتعلم كيف نجعل الأرقام **تفهم المعاني**،
> من خلال تقنيات أكثر ذكاءً مثل **TF-IDF** و **Word2Vec**.
> استعد لأن القادم هو بداية السحر الحقيقي في الـ NLP! ✨

> 🇬🇧
> Now that we’ve learned how to represent text with simple numbers using BoW,
> it’s time to move forward and make those numbers **understand meaning**.
> Next, we’ll explore smarter methods like **TF-IDF** and **Word2Vec** —
> where mathematics meets language magic. ✨

---

### 🇸🇦 مقدمة

في هذا القسم سنقوم بتطبيق **Bag of Words (BoW)** عمليًا باستخدام بايثون.
سنحوّل النصوص إلى تمثيلات رقمية تُظهر **وجود الكلمات وتكرارها**، دون النظر إلى معناها.
رغم أن هذا الأسلوب قديم نوعًا ما في 2025، إلا أنه لا يزال مفيدًا جدًا في بعض المهام .

---

### 🇬🇧 Introduction

In this section, we’ll implement **Bag of Words (BoW)** in Python.
We’ll convert text into numerical vectors showing **the presence and frequency of words**,
without understanding their meaning.

Although BoW is considered a **classic** method by 2025, it’s still useful.
---

In [12]:
import nltk 
import re
import numpy
from nltk.corpus import stopwords

# Function: clean and split sentence into words
def word_extraction(sentence):
    # Remove punctuation and split text
    words = re.sub("[^\w]", " ", sentence).split()
    # Convert to lowercase and remove English stopwords
    cleaned_text = [w.lower() for w in words if w not in stopwords.words('english')]
    return cleaned_text

# Function: create sorted list of unique words (vocabulary)
def tokenize(sentences):
    words = []
    for sentence in sentences:
        w = word_extraction(sentence)
        words.extend(w)
    # Remove duplicates and sort alphabetically
    words = sorted(list(set(words)))
    return words


### 🇸🇦 الشرح

الدالة `word_extraction()` تنظّف النص من الرموز وتحذف كلمات التوقف مثل “the”, “and”, “is”.
أما `tokenize()` فتنشئ **قائمة بالكلمات الفريدة** التي سنبني عليها نموذج الـ BoW.

---

### 🇬🇧 Explanation

`word_extraction()` cleans the text by removing punctuation and stopwords.
`tokenize()` creates a **sorted list of unique words**, forming the vocabulary for the BoW model.

---

In [13]:
# Example text
text = "Mary and Samantha arrived at the bus station early but waited until noon for the bus"

# Test cleaning function
word_extraction(text)

# Test tokenization
tokenize(word_extraction(text))


['arrived', 'bus', 'early', 'mary', 'noon', 'samantha', 'station', 'waited']

In [14]:
# Function: Generate Bag of Words manually
def generate_bow(allsentences):
    vocab = tokenize(allsentences)
    print("Word List for Document \n{0} \n".format(vocab))
    
    for sentence in allsentences:
        words = word_extraction(sentence)
        bag_vector = numpy.zeros(len(vocab))  # Create a vector of zeros
        
        # Count each word's presence in the sentence
        for w in words:
            for i, word in enumerate(vocab):
                if word == w:
                    bag_vector[i] += 1
        
        print("{0}\n{1}\n".format(sentence, numpy.array(bag_vector)))
        print('---------------------------------------------------')


### 🇸🇦 الفكرة

كل جملة تتحول إلى **متجه من الأرقام** يعبّر عن الكلمات التي تحتويها،
بدون أي فهم للمعنى أو الترتيب.

لكن لاحظ 👇
BoW لا يدرك الفرق بين:

> “جميل” و “مشرق”
> ولا يعرف أن “اليوم ممطر” تعني شيئًا عكس “اليوم مشمس”.

🔹 في 2025، نادراً ما يُستخدم BoW لوحده،
بل يُستخدم مع خوارزميات أبسط أو لتحليل سريع قبل استخدام تقنيات مثل **Word2Vec** أو **Transformer Models**.

---

### 🇬🇧 Concept

Each sentence becomes a **vector of numbers**, representing which words appear in it —
without understanding meaning or word order.

BoW doesn’t recognize relationships like:

> “nice” ≈ “beautiful”
> “rainy” ≠ “sunny”

🔹 In 2025, BoW is rarely used alone.
It’s mainly used for **quick text analysis** or in **lightweight NLP models** before moving to more advanced embeddings like **Word2Vec** or **BERT**.

---

In [15]:
allsentences = [
    "Joe waited for the train",
    "The train was late",
    "Mary and Samantha took the bus",
    "I looked for Mary and Samantha at the bus station",
    "Mary and Samantha arrived at the bus station early but waited until noon for the bus"
]

# Generate BoW representation for each sentence
generate_bow(allsentences)


Word List for Document 
['arrived', 'bus', 'early', 'i', 'joe', 'late', 'looked', 'mary', 'noon', 'samantha', 'station', 'the', 'took', 'train', 'waited'] 

Joe waited for the train
[0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 1. 1.]

---------------------------------------------------
The train was late
[0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 1. 0. 1. 0.]

---------------------------------------------------
Mary and Samantha took the bus
[0. 1. 0. 0. 0. 0. 0. 1. 0. 1. 0. 0. 1. 0. 0.]

---------------------------------------------------
I looked for Mary and Samantha at the bus station
[0. 1. 0. 1. 0. 0. 1. 1. 0. 1. 1. 0. 0. 0. 0.]

---------------------------------------------------
Mary and Samantha arrived at the bus station early but waited until noon for the bus
[1. 2. 1. 0. 0. 0. 0. 1. 1. 1. 1. 0. 0. 0. 1.]

---------------------------------------------------


In [16]:
from sklearn.feature_extraction.text import CountVectorizer

# Create vectorizer
vectorizer = CountVectorizer()

# Transform sentences into numerical vectors
X = vectorizer.fit_transform(allsentences)

# Display results
for i in range(len(allsentences)):
    print(allsentences[i])
    print(list(X.toarray()[i]))
    print('---------------------------------------------------')


Joe waited for the train
[0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0]
---------------------------------------------------
The train was late
[0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1]
---------------------------------------------------
Mary and Samantha took the bus
[1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0]
---------------------------------------------------
I looked for Mary and Samantha at the bus station
[1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0]
---------------------------------------------------
Mary and Samantha arrived at the bus station early but waited until noon for the bus
[1, 1, 1, 2, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 2, 0, 0, 1, 1, 0]
---------------------------------------------------


### 🇸🇦 الشرح

بدلاً من بناء BoW يدويًا، توفر مكتبة **scikit-learn** أداة `CountVectorizer`
التي تقوم بكل العمليات تلقائيًا — تنظيف، تجزئة، وعدّ الكلمات.

لكن BoW يظل **بسيطًا وفعالًا** مع النصوص القصيرة أو التطبيقات التعليمية.

---

### 🇬🇧 Explanation

Instead of manually building BoW, the **scikit-learn** library provides `CountVectorizer`,
which automates cleaning, tokenizing, and counting.

BoW remains **simple and effective** for small-scale or educational NLP tasks.

---

* BoW يعمل **بشكل ممتاز مع اللغة العربية** لأن آليته تعتمد على **تكرار الكلمة** وليس معناها.
* BoW performs **very well with Arabic** since it focuses on **word occurrence**, not meaning.


In [17]:
text = "ذهب محمد و شريف الي المدرسة هذا اليوم"
word_extraction(text)
tokenize(word_extraction(text))
generate_bow(word_extraction(text))


Word List for Document 
['المدرسة', 'الي', 'اليوم', 'ذهب', 'شريف', 'محمد', 'هذا', 'و'] 

ذهب
[0. 0. 0. 1. 0. 0. 0. 0.]

---------------------------------------------------
محمد
[0. 0. 0. 0. 0. 1. 0. 0.]

---------------------------------------------------
و
[0. 0. 0. 0. 0. 0. 0. 1.]

---------------------------------------------------
شريف
[0. 0. 0. 0. 1. 0. 0. 0.]

---------------------------------------------------
الي
[0. 1. 0. 0. 0. 0. 0. 0.]

---------------------------------------------------
المدرسة
[1. 0. 0. 0. 0. 0. 0. 0.]

---------------------------------------------------
هذا
[0. 0. 0. 0. 0. 0. 1. 0.]

---------------------------------------------------
اليوم
[0. 0. 1. 0. 0. 0. 0. 0.]

---------------------------------------------------


In [18]:
allsentences = [
    "ذهب محمد الي الجامعة",
    "نجحت مني في الاختبار",
    "تمكن محمد من السفر هذا اليوم",
    "اعتقد ان جورج سيتصل بنا اليوم",
    "لا اظن ان الاختبار سيكون سهلا"
]

generate_bow(allsentences)


Word List for Document 
['اظن', 'اعتقد', 'الاختبار', 'الجامعة', 'السفر', 'الي', 'اليوم', 'ان', 'بنا', 'تمكن', 'جورج', 'ذهب', 'سهلا', 'سيتصل', 'سيكون', 'في', 'لا', 'محمد', 'من', 'مني', 'نجحت', 'هذا'] 

ذهب محمد الي الجامعة
[0. 0. 0. 1. 0. 1. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]

---------------------------------------------------
نجحت مني في الاختبار
[0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 1. 1. 0.]

---------------------------------------------------
تمكن محمد من السفر هذا اليوم
[0. 0. 0. 0. 1. 0. 1. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 1. 1. 0. 0. 1.]

---------------------------------------------------
اعتقد ان جورج سيتصل بنا اليوم
[0. 1. 0. 0. 0. 0. 1. 1. 1. 0. 1. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]

---------------------------------------------------
لا اظن ان الاختبار سيكون سهلا
[1. 0. 1. 0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 0. 1. 0. 1. 0. 0. 0. 0. 0.]

---------------------------------------------------


---

# 🧮 الجزء الرابع: 🇸🇦 **TF-IDF** | 🇬🇧 **Term Frequency – Inverse Document Frequency**

---

## 🇸🇦 الفكرة الأساسية

طريقة **TF-IDF** تُعد من أهم الأدوات الإحصائية في معالجة اللغة الطبيعية **(NLP)**،
وتُستخدم بكثرة في مهام مثل **تصنيف النصوص (Text Classification)** و**تحليل المشاعر (Sentiment Analysis)**.

هي تطوير ذكي لفكرة **Bag of Words (BoW)**:

* BoW كان يهتم فقط **بوجود الكلمة وتكرارها**.
* TF-IDF يضيف **عاملًا ذكيًا** يوازن بين **تكرار الكلمة داخل النص** و**ندرتها عبر مجموعة النصوص**.

الهدف هو التمييز بين **الكلمات المميزة فعلًا** و**الكلمات العامة عديمة الفائدة** مثل "هذا" أو "الذي".

---

## 🇬🇧 Core Concept

**TF-IDF** is a powerful statistical method in **Natural Language Processing (NLP)**.
It’s widely used in **text classification** and **sentiment analysis**.

It builds upon the **Bag of Words** model:

* BoW only counts word occurrences.
* TF-IDF adds intelligence by balancing **word frequency within a document** and **word rarity across documents**.

The goal is to highlight **unique, meaningful words** and down-weight **common filler words** like “the” or “is”.

---

## 🧱 الفرق بين TF-IDF و BoW | Difference from BoW

| BoW                                   | TF-IDF                           |
| ------------------------------------- | -------------------------------- |
| يحسب وجود الكلمة فقط                  | يحسب وجودها × أهميتها            |
| لا يفرّق بين الكلمات الشائعة والنادرة | يعطي وزنًا أكبر للكلمات النادرة  |
| يعتمد على التكرار الخام               | يعتمد على التكرار النسبي والندرة |
| مناسب للإحصاء البسيط فقط              | مناسب للفهم الإحصائي الأعمق      |

---

## ⚙️ 🇸🇦 كيف بدأ التطور؟

تخيل أنك تحلل **روايات شكسبير** وتحصي أسماء الأبطال.
باستخدام BoW، تعرف فقط إن كانت كلمة “قيصر” موجودة أم لا.
لكن TF-IDF يذهب أبعد من ذلك — فهو يعرف أنها **تكررت أكثر بـ150 مرة من "ميرسي"**،
ويفهم أن “قيصر” **كلمة محورية في الرواية**.

لكن لاحظ 👇
BoW وTF يعانيان من مشكلة: **الترتيب لا يُؤخذ بالحسبان.**
الجملتان:

> “فرنسا دولة غنية بينما ألمانيا يتراجع اقتصادها.”
> “ألمانيا دولة غنية بينما فرنسا يتراجع اقتصادها.”
> لهما نفس قيمة BoW، رغم اختلاف المعنى.
> ولذلك جاء تطوير “الترقيم الموضعي (Positional Encoding)” لاحقًا لعلاج هذا النقص،
> لكنّه موضوع لاحق خارج نطاق TF-IDF.

---

## 🇬🇧 Evolution Example

Imagine analyzing **Shakespeare’s novels**:
BoW tells you if “Caesar” appears or not,
but TF-IDF shows that “Caesar” appeared **150× more often** than “Mercy”,
so it must be **more central to the story**.

However, word order is ignored —
so “France is rich but Germany’s economy is falling”
is treated the same as the reverse version.
That’s one limitation of TF-only methods, later improved by **positional encoding** techniques.

---

## 🧮 المكونات الأساسية | Components

| الرمز        | المصطلح                      | الشرح                                                           |
| ------------ | ---------------------------- | --------------------------------------------------------------- |
| **TF**       | *Term Frequency*             | عدد مرات تكرار الكلمة داخل المستند ÷ العدد الكلي لكلمات المستند |
| **IDF**      | *Inverse Document Frequency* | مدى ندرة الكلمة عبر جميع المستندات                              |
| **TF × IDF** | الوزن النهائي                | قيمة تُمثل أهمية الكلمة داخل المستند مقارنةً ببقية المستندات    |

---

## 🧠 🇸🇦 لماذا نستخدم اللوغاريتم؟

لو ظهرت كلمة 20 مرة في مستند، هذا لا يعني أن المستند أهم 20 مرة من آخر ظهرت فيه مرة واحدة.
نستخدم اللوغاريتم لتقليل الفروق الكبيرة بين التكرارات العالية.
فتصبح العلاقة غير خطية أكثر واقعية:

* مرة واحدة → 1
* مرتين → 1.3
* 10 مرات → 2
* 1000 مرة → 4

بالتالي، TF يعكس **مدى تكرار الكلمة داخل النص**،
لكن **IDF يعكس مدى ندرتها عبر النصوص**.

---

## 🇬🇧 Why a Logarithmic Scale?

Because frequency grows too fast linearly.
A word appearing 20 times doesn’t mean it’s 20× more important.
Using logarithms compresses large frequencies:
1 → 1, 2 → 1.3, 10 → 2, 1000 → 4.

Thus, TF captures **within-document frequency**,
while IDF captures **across-document rarity**.

---

## 📐 🇸🇦 حساب IDF خطوة بخطوة

نحسب رقمين:

* **N:** عدد جميع المستندات.
* **df:** عدد المستندات التي تحتوي الكلمة.

[
IDF = \log\left(\frac{N}{df}\right)
]

كلما كانت الكلمة نادرة → زادت قيمة IDF
كلما كانت شائعة → قلت قيمتها.

مثلًا:

* ظهرت في 10 مقالات فقط من مليون → log(1,000,000/10) = 5
* ظهرت في نصفها → log(2) = 0.3
* ظهرت في جميعها → log(1) = 0 → أي غير مفيدة إحصائيًا.

---

## 🇬🇧 Step-by-Step IDF Calculation

Two values:

* **N:** Total number of documents
* **df:** Number of documents containing the word

[
IDF = \log\left(\frac{N}{df}\right)
]

A word appearing in all documents → IDF = 0
Rare word → higher IDF, hence more informative.

---

## 🧩 🇸🇦 مثال عملي مبسّط

افترض كلمة **"Trump"** ظهرت 50 مرة في مقال يحتوي 1000 كلمة
وذكرت في 10,000 من أصل مليون مقال.

إذن:

* TF = 50 ÷ 1000 = 0.05
* IDF = log(1,000,000 ÷ 10,000) = log(100) = 2
  → TF-IDF = 0.05 × 2 = **0.1**

تُظهر هذه القيمة أن الكلمة مميزة نسبيًا داخل المقال.

---

## 🇬🇧 Example

If the word “Trump” appears 50 times in a 1000-word article
and in 10,000 out of 1,000,000 articles overall:

* TF = 0.05
* IDF = log(100) = 2
* TF-IDF = 0.1

The higher the score, the more representative the word is for that document.

---

## ⚖️ 🇸🇦 الاستخدامات العملية

* تقييم أهمية الكلمات داخل النصوص.
* استخراج الكلمات المفتاحية.
* ترتيب نتائج البحث (Search Ranking).
* تحليل المشاعر وتصنيف النصوص.
* تجهيز البيانات لخوارزميات تعلم الآلة.

وفي 2025، لا تزال **TF-IDF** تُستخدم على نطاق واسع لأنها **بسيطة وسريعة وفعالة**،
خصوصًا مع النصوص العربية حيث **تكرار الكلمات أكثر أهمية من معناها** أحيانًا.

---

## 🇬🇧 Real-World Uses (as of 2025)

* Keyword extraction
* Search engine ranking
* Sentiment analysis
* Document classification
* Feature extraction for ML models

In 2025, TF-IDF remains popular for **Arabic NLP**,
since it values **word occurrence statistics** over deep semantics —
a great match for tasks like **NER** and **POS tagging** foundations.

---

## 🧩 🇸🇦 العلاقة مع نماذج 2025 الحديثة

في السنوات الأخيرة، لم تختفِ TF-IDF بل تم **دمجها** مع تمثيلات أعمق مثل **AraBERT** و**FastText**
لإنشاء **نموذج هجين Hybrid Model** يجمع بين:

* قوة TF-IDF الإحصائية
* وفهم الـ Embeddings الدلالي.

وهذا مفيد جدًا في المجالات العربية التي تحتوي على نصوص مختلطة بين اللغة الفصحى والعامية.

---

## 🇬🇧 TF-IDF in Modern (2025) Systems

In recent years, TF-IDF hasn’t disappeared — it evolved.
It’s now combined with **semantic embeddings** (like AraBERT or FastText)
to create **hybrid systems** that blend statistical precision with contextual understanding.

This is especially effective for **Arabic NLP**,
where meaning often depends on context and morphology.

---

## 📚 🇸🇦 المعادلة النهائية

[
TF\text{-}IDF(w, d) = TF(w, d) \times \log\left(\frac{N}{df(w)}\right)
]

كلما:

* زاد تكرار الكلمة في المستند
* وقلّ انتشارها في بقية المستندات
  ⬅️ زاد وزنها وأهميتها.

---

## 🇬🇧 Final Formula

[
TF\text{-}IDF(w, d) = TF(w, d) \times \log\left(\frac{N}{df(w)}\right)
]

A word is more important when:

* It appears frequently in one document
* But rarely across others.

---

## 💬 🇸🇦 خلاصة بسيطة | 🇬🇧 Quick Summary

| 🇸🇦                                       | 🇬🇧                                                   |
| ------------------------------------------ | ------------------------------------------------------ |
| TF-IDF تطوير ذكي لفكرة BoW                 | TF-IDF is an advanced form of BoW                      |
| يعطي وزنًا للكلمات النادرة والمميزة        | Gives higher weight to rare, informative words         |
| يقلل من أثر الكلمات الشائعة                | Reduces influence of frequent, generic words           |
| ما زال فعالًا جدًا خصوصًا مع اللغة العربية | Still effective — especially for Arabic text           |
| يُستخدم كثيرًا قبل التدريب في خوارزميات ML | Often used for ML preprocessing and feature extraction |

---

---

## 🧠 🇸🇦 مقدمة عملية | 🇬🇧 Practical Introduction

الآن سنقوم بتطبيق طريقة **TF-IDF** يدويًا أولًا لفهم المبدأ الرياضي،
ثم نستخدم مكتبة **Scikit-learn** لتطبيقها مباشرة.
الهدف هو فهم كيف يتم حساب:

* **TF** (عدد تكرارات الكلمة في النص)
* **IDF** (مدى ندرة الكلمة عبر المستندات)
* ثم **TF × IDF** (الأهمية النهائية للكلمة)

---

In [19]:
import pandas as pd

# 🏁 Example Documents
# 🇬🇧 Two short documents for comparison
# 🇸🇦 مثال بسيط على جملتين لتحليلها باستخدام TF-IDF
documentA = 'the man went out for a walk'
documentB = 'the children sat around the fire'

# Split each document into words (tokenization)
bagOfWordsA = documentA.split(' ')
bagOfWordsB = documentB.split(' ')
bagOfWordsA


['the', 'man', 'went', 'out', 'for', 'a', 'walk']

---

### 🇸🇦 الخطوة الأولى: إنشاء قائمة بالكلمات الفريدة

### 🇬🇧 Step 1: Create the unique vocabulary

In [20]:
uniqueWords = set(bagOfWordsA).union(set(bagOfWordsB))
uniqueWords


{'a',
 'around',
 'children',
 'fire',
 'for',
 'man',
 'out',
 'sat',
 'the',
 'walk',
 'went'}

---

### 🇸🇦 الخطوة الثانية: حساب عدد مرات ظهور كل كلمة

### 🇬🇧 Step 2: Count occurrences of each word

In [21]:
numOfWordsA = dict.fromkeys(uniqueWords, 0)
for word in bagOfWordsA:
    numOfWordsA[word] += 1

numOfWordsB = dict.fromkeys(uniqueWords, 0)
for word in bagOfWordsB:
    numOfWordsB[word] += 1


---

### 🇸🇦 الخطوة الثالثة: حساب TF

**TF = عدد مرات تكرار الكلمة ÷ عدد الكلمات الكلي في المستند**

### 🇬🇧 Step 3: Compute Term Frequency (TF)

**TF = (count of a word in the document) / (total number of words)**

In [22]:
def computeTF(wordDict, bagOfWords):
    tfDict = {}
    bagOfWordsCount = len(bagOfWords)
    for word, count in wordDict.items():
        tfDict[word] = count / float(bagOfWordsCount)
    return tfDict

tfA = computeTF(numOfWordsA, bagOfWordsA)
tfB = computeTF(numOfWordsB, bagOfWordsB)


📘 **ملاحظة:** TF يُظهر مدى تكرار الكلمة في النص،
لكن لا يأخذ في الاعتبار مدى شيوعها في بقية المستندات.

---

### 🇸🇦 الخطوة الرابعة: حساب IDF

**IDF = log(عدد كل المستندات ÷ عدد المستندات التي تحتوي الكلمة)**

### 🇬🇧 Step 4: Compute Inverse Document Frequency (IDF)

**IDF = log(N / df)**

In [23]:
import math

def computeIDF(documents):
    N = len(documents)  # total number of documents
    idfDict = dict.fromkeys(documents[0].keys(), 0)
    for document in documents:
        for word, val in document.items():
            if val > 0:
                idfDict[word] += 1

    for word, val in idfDict.items():
        idfDict[word] = math.log(N / float(val))
    return idfDict

idfs = computeIDF([numOfWordsA, numOfWordsB])
idfs


{'the': 0.0,
 'fire': 0.6931471805599453,
 'around': 0.6931471805599453,
 'sat': 0.6931471805599453,
 'children': 0.6931471805599453,
 'man': 0.6931471805599453,
 'went': 0.6931471805599453,
 'out': 0.6931471805599453,
 'walk': 0.6931471805599453,
 'a': 0.6931471805599453,
 'for': 0.6931471805599453}

📘 **ملاحظة مهمة:**

كلما ظهرت الكلمة في عدد كبير من النصوص، قلّت قيمتها.
وهذا ما يجعل TF-IDF أكثر ذكاءً من BoW.

---

### 🇸🇦 الخطوة الخامسة: حساب TF × IDF

### 🇬🇧 Step 5: Compute TF-IDF

In [24]:
def computeTFIDF(tfBagOfWords, idfs):
    tfidf = {}
    for word, val in tfBagOfWords.items():
        tfidf[word] = val * idfs[word]
    return tfidf

tfidfA = computeTFIDF(tfA, idfs)
tfidfB = computeTFIDF(tfB, idfs)

df = pd.DataFrame([tfidfA, tfidfB])
df.head()


the      fire    around       sat  children       man      went       out  \
0  0.0  0.000000  0.000000  0.000000  0.000000  0.099021  0.099021  0.099021   
1  0.0  0.115525  0.115525  0.115525  0.115525  0.000000  0.000000  0.000000   

       walk         a       for  
0  0.099021  0.099021  0.099021  
1  0.000000  0.000000  0.000000

---

## ⚙️ 🇸🇦 التطبيق باستخدام مكتبة sklearn

### 🇬🇧 Using Scikit-learn’s Built-in TF-IDF Vectorizer

In [25]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer()
vectors = vectorizer.fit_transform([documentA, documentB])

feature_names = vectorizer.get_feature_names_out()
dense = vectors.todense()
df = pd.DataFrame(dense, columns=feature_names)
df.head()


around  children      fire      for      man      out       sat  \
0  0.000000  0.000000  0.000000  0.42616  0.42616  0.42616  0.000000   
1  0.407401  0.407401  0.407401  0.00000  0.00000  0.00000  0.407401   

        the     walk     went  
0  0.303216  0.42616  0.42616  
1  0.579739  0.00000  0.00000

---

## 📘 🇸🇦 مثال آخر باللغة العربية | 🇬🇧 Arabic Example


In [26]:
documentA = 'ذهب محمد الي الجامعة ليدرس الفيزياء و الكيمياء'
documentB = 'ذاكرت مني الرياضيات و الفيزياء في الجامعة'

# Tokenize (split words)
bagOfWordsA = documentA.split(' ')
bagOfWordsB = documentB.split(' ')

# Create vocabulary
uniqueWords = set(bagOfWordsA).union(set(bagOfWordsB))

# Count word frequencies
numOfWordsA = dict.fromkeys(uniqueWords, 0)
numOfWordsB = dict.fromkeys(uniqueWords, 0)
for word in bagOfWordsA: numOfWordsA[word] += 1
for word in bagOfWordsB: numOfWordsB[word] += 1

# Compute TF
tfA = computeTF(numOfWordsA, bagOfWordsA)
tfB = computeTF(numOfWordsB, bagOfWordsB)

# Compute IDF
idfs = computeIDF([numOfWordsA, numOfWordsB])

# Compute TF-IDF
tfidfA = computeTFIDF(tfA, idfs)
tfidfB = computeTFIDF(tfB, idfs)

pd.DataFrame([tfidfA, tfidfB])


الي  الجامعة  الرياضيات      محمد       مني       ذهب        في  \
0  0.086643      0.0   0.000000  0.086643  0.000000  0.086643  0.000000   
1  0.000000      0.0   0.099021  0.000000  0.099021  0.000000  0.099021   

   الفيزياء    و  الكيمياء     ليدرس     ذاكرت  
0       0.0  0.0  0.086643  0.086643  0.000000  
1       0.0  0.0  0.000000  0.000000  0.099021

📙 **ملاحظة مهمة:**
TF-IDF تعمل بكفاءة عالية مع اللغة العربية،
لأنها تهتم **بتكرار الكلمة وليس معناها**،
وهذا يجعلها مفيدة جدًا في مهام مثل:

* **تحليل النصوص العربية (Text Mining)**
* **تصنيف المقالات**
* **التعرف على الكلمات المهمة (Keyword Extraction)**

---



## 🚀 🇸🇦 باستخدام Scikit-learn مع اللغة العربية

### 🇬🇧 Arabic Text TF-IDF using sklearn

In [27]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer()
vectors = vectorizer.fit_transform([documentA, documentB])

feature_names = vectorizer.get_feature_names_out()
dense = vectors.todense()
pd.DataFrame(dense, columns=feature_names)


الجامعة  الرياضيات  الفيزياء  الكيمياء       الي     ذاكرت       ذهب  \
0  0.29017   0.000000   0.29017  0.407824  0.407824  0.000000  0.407824   
1  0.31780   0.446656   0.31780  0.000000  0.000000  0.446656  0.000000   

         في     ليدرس      محمد       مني  
0  0.000000  0.407824  0.407824  0.000000  
1  0.446656  0.000000  0.000000  0.446656

---

## 🧠 🇸🇦 ملاحظات مهمة (تحديث 2025) | 🇬🇧 Key Notes (as of 2025)

| 🇸🇦 الملاحظة                                    | 🇬🇧 Note                                                  |
| ------------------------------------------------ | ---------------------------------------------------------- |
| TF-IDF لا يفهم المعنى أو الترتيب                 | TF-IDF doesn’t capture semantic meaning or order           |
| لكنه ممتاز في استخراج الكلمات المهمة             | Great for identifying significant terms                    |
| يُستخدم كأساس قبل نماذج أعمق مثل BERT            | Often used as preprocessing before deep models             |
| يدعم العربية بدقة عالية خصوصًا في النصوص الرسمية | Works very well with Modern Standard Arabic                |
| في 2025 يُستخدم مدموجًا مع Word Embeddings       | In 2025, often combined with embeddings for hybrid systems |

---

---
**ملاحظات بحثت عنها لزيادة الفهم**


### 🧱 أولًا: ليش TF-IDF ما يهتم بالترتيب؟

لأن فكرته الأصلية **إحصائية بحتة**،
يعني ما يفهم المعنى، هو بس **يعدّ الكلمات** كم مرة ظهرت.

> هو يشوف النص مثل “كيس كلمات” (Bag of Words)،
> ما يهمه وين الكلمة جات، المهم كم مرة جات.

📘 مثال:

> “الولد أكل التفاحة”
> “التفاحة أكل الولد”

بالنسبة لـ TF-IDF:
نفس الكلمات بالضبط → نفس القيم → نفس التمثيل،
مع إنه المعنى مختلف تمامًا.

السبب؟ لأنه ببساطة **ما صُمم لفهم اللغة**،
بل فقط لحساب **أهمية الكلمات** داخل النصوص.

---

### 🧠 ثانيًا: ليه العلماء ما خلوه يهتم بالترتيب؟

لأن هدفهم وقتها (قبل ظهور الذكاء العميق)
كان بس **تمييز النصوص وتصنيفها بسرعة**،
مو فهم المعنى الكامل.
الترتيب يحتاج نماذج أعقد بكثيير (مثل الشبكات العصبية)،
وهذا كان صعب أو مكلف وقتها.

---

### 🚀 ثالثًا: هل النماذج الجديدة تهتم بالترتيب؟

نعم ✅
النماذج الحديثة مثل:

> **Word2Vec, BERT, AraBERT, GPT…**

تهتم **بالسياق والترتيب والمعنى**.
يعني تعرف الفرق بين:

> “أكل الولد التفاحة” 🍎
> “التفاحة أكلت الولد” 😨

لأنها تفهم أن المعنى تغيّر حسب ترتيب الكلمات.

---

📍 **الخلاصة:**

* **TF-IDF:** يعدّ الكلمات فقط → ما يهمه الترتيب.
* **النماذج الحديثة:** تفهم الترتيب والمعنى والسياق.

---



### 💡 **TF-IDF قوي لكنه محدود**

نعم، تقدر تبني عليه تطبيقات وتدرب نماذج،
لكن **دقته محدودة** في فهم المعنى.

خلني أوضحها ببساطة:

---

### ✅ متى TF-IDF ممتاز

* لما الهدف **تمييز أو تصنيف النصوص** (مثل معرفة هل التغريدة إيجابية أو سلبية).
* لما البيانات كثيرة والكلمات واضحة ومميزة.
* لما ما تحتاج تفهم **المعنى الدقيق أو ترتيب الجمل**.

📘 مثال:

> تبي تصنف مراجعات منتجات (جيدة / سيئة)
> TF-IDF يعطيك نتائج حلوة وسريعة وفعّالة 💪

---

### ⚠️ متى يصير ضعيف

* لما النصوص معقدة أو فيها **معاني ضمنية وسياق**.
* لما الترتيب يغيّر المعنى (مثل “ما أحب الكذب” ≠ “أحب الكذب”).
* لما تتعامل مع **اللغة الطبيعية المتنوعة** (مثل العربية العامية أو النصوص الطويلة).

في هالحالات، TF-IDF ما يفهم “النية” أو “السياق”،
فنتائجه تكون أقل دقة مقارنة بالنماذج الحديثة.

---


### 🧠 مثال بسيط

عندك بيانات فيها:

| النص                      | التصنيف |
| ------------------------- | ------- |
| هذا المطعم رائع جدًا      | إيجابي  |
| الأكل لذيذ والخدمة ممتازة | إيجابي  |
| الأكل سيئ جدًا            | سلبي    |
| الخدمة بطيئة              | سلبي    |

لما نحول النصوص إلى **TF-IDF**،
النموذج بيلاحظ إحصائيًا أن:

* “رائع” و“ممتاز” غالبًا تظهر في نصوص إيجابية ✅
* “سيئ” و“بطيئة” تظهر في نصوص سلبية ❌

فالنموذج (مثل Logistic Regression أو SVM)
يتعلم العلاقة الإحصائية بين **الكلمات والتصنيف**،
بدون ما يفهم معنى الكلمة فعليًا.

---

### ⚠️ إذًا وش العيب هنا؟

لو جته جملة مثل:

> “المطعم كان رائع قبل سنة، بس صار سيئ الآن”
> TF-IDF ممكن يحتار 😅
> لأن فيها “رائع” (إيجابية) و“سيئ” (سلبية)،
> وهو ما يعرف السياق ولا الترتيب،
> فما يقدر يفهم أن المعنى العام سلبي.

---

### 🧩 بينما النماذج الحديثة (BERT، GPT...)

هذي **تفهم المعنى والسياق**،
فتعرف أن الجملة تقول إن التجربة ساءت مع الوقت،
وتصنفها “سلبية” بسهولة ✅

---


---

## 🧠 🇸🇦 الجزء الرابع: Word2Vec (أو Vec2Word)

## 🇬🇧 Section 7: Word2Vec (Vec2Word)

---

### 🪄 🇸🇦 الفكرة العامة

**Word2Vec** (أو **Vec2Word**) هي خوارزمية مبنية على **شبكة عصبية بسيطة مكونة من طبقتين**.
هدفها الأساسي هو تحويل الكلمات إلى **تمثيلات رقمية (Vectors)** بحيث:

* الكلمات المتشابهة في المعنى تكون قريبة رياضيًا.
* الكلمات المختلفة تكون بعيدة.

👩‍💻 مثلًا بدل ما تكون الكلمة “سلام” مجرد نص،
تصبح متجهًا رقميًا مثل:
`سلام → [0.52, 0.31, -0.22, ...]`
وهذا يسمح للنموذج أن “يفهم” العلاقات مثل:

> الرجل - صبي ≈ المرأة - فتاة
> king - man + woman ≈ queen 👑

---

### 🇬🇧 The General Idea

**Word2Vec** (also called **Vec2Word**) is a **two-layer neural network** designed to transform words into **numerical representations (vectors)**.
Its goal is to capture **semantic meaning** mathematically so that:

* Words with similar meanings are **closer together** in vector space.
* Words with different meanings are **farther apart**.

👩‍💻 For example, instead of treating “peace” as just text, it becomes:
`peace → [0.52, 0.31, -0.22, ...]`
This allows the model to understand analogies like:

> man - boy ≈ woman - girl
> king - man + woman ≈ queen 👑

This process is called **semantic embedding** — representing meaning as numbers.

---

## ⚙️ 🇸🇦 كيف تشتغل الفكرة؟

## 🇬🇧 How It Works

النموذج يتعلم من خلال فكرة بسيطة:

> “الكلمة تُعرف من الكلمات التي حولها.”

📘 مثال:
في الجملة “الولد ذهب إلى المدرسة”،
كلمة **ذهب** تظهر دائمًا بجانب “الولد” و“المدرسة” و“إلى”،
فيتعلم النموذج أن “ذهب” ترتبط بالسياق الحركي أو الفعل.

---

📘 Example:
In the sentence “The boy went to school,”
the word **went** often appears near “boy,” “to,” and “school.”
So the model learns that **went** is semantically related to motion or actions.

---

## 🔢 🇸🇦 أنواع Word2Vec

## 🇬🇧 The Two Main Architectures

Word2Vec يتكون من نموذجين أساسيين:

1. **CBOW – Continuous Bag of Words**
2. **Skip-Gram**

---

### 🧩 🇸🇦 أولًا: نموذج CBOW

**(Continuous Bag of Words)**

#### 🇸🇦 الفكرة

نموذج **CBOW** يتعلم من خلال **توقع الكلمة المفقودة** في الجملة اعتمادًا على **الكلمات المحيطة بها**.

📘 مثال:

> “الطالب … إلى الجامعة كل يوم.”
> النموذج يحاول أن يتنبأ بالكلمة الناقصة “يذهب” بناءً على الكلمات التي قبلها وبعدها.

---

#### 🇬🇧 The Idea

The **CBOW** model learns by **predicting a missing word** based on the **surrounding words** in a sentence.

📘 Example:

> “The student … to the university every day.”
> The model predicts the missing word “goes” from its context words.

---

### 🔢 🇸🇦 طريقة العمل بالتفصيل

### 🇬🇧 Detailed Working Steps

1. نختار **نافذة (window)** من الكلمات حول الكلمة المستهدفة (مثل كلمتين قبل وبعد).
2. نحول كل كلمة إلى تمثيل رقمي (embedding vector).
3. نحسب المتوسط لهذه التمثيلات.
4. ندخله في **الطبقة الخفية (hidden layer)**.
5. نستخدم **Softmax** في الطبقة الأخيرة لتوقع الكلمة المفقودة.

📘 مثال:

> “the ___ sat on the mat”
> الكلمات المحيطة → {the, sat, on, the, mat}
> النموذج يتعلم أن الكلمة المفقودة غالبًا “cat”.

---

📘 Example (English):

1. Choose a **window** of words around the target word (e.g., two before and two after).
2. Convert each into a numerical embedding.
3. Compute the **average** of those embeddings.
4. Pass it through a **hidden layer**.
5. Use **Softmax** to predict the missing word.

> “the ___ sat on the mat”
> Context words → {the, sat, on, the, mat}
> The model learns that the missing word is probably “cat.”

---

### 💡 🇸🇦 الاستخدامات الواقعية

### 🇬🇧 Real-World Applications

أحد أشهر التطبيقات هو:

> **Spinning Articles** 🌀
> وهو إعادة صياغة النصوص تلقائيًا باستخدام كلمات مرادفة للحفاظ على نفس المعنى.

🔹 الطريقة:

1. نحذف بعض الكلمات من النص.
2. CBOW يتنبأ بالكلمات المناسبة من السياق.
3. الناتج: نص جديد بنفس المعنى ولكن بصياغة مختلفة.

---

One popular application is:

> **Article Spinning** 🌀
> It rewrites existing text automatically by replacing words with synonyms while preserving meaning.

🔹 Process:

1. Some words are removed from the text.
2. CBOW predicts suitable replacements from context.
3. The result is a rephrased text with the same meaning.

---

### 🧠 🇸🇦 ثانيًا: نموذج Skip-Gram

### 🇬🇧 Second: Skip-Gram Model

#### 🇸🇦 الفكرة

هو **عكس CBOW** تقريبًا.
بدل ما يتنبأ بالكلمة المفقودة من السياق،
يتنبأ **بالكلمات المحيطة** اعتمادًا على كلمة واحدة.

📘 مثال:

> الكلمة “طبيب” → يتوقع الكلمات القريبة منها مثل “مريض”، “مستشفى”، “دواء”.

---

#### 🇬🇧 The Idea

It works almost **opposite to CBOW**.
Instead of predicting a missing word from context,
it predicts the **context words** given a single word.

📘 Example:

> Input word: “doctor” → predicts words like “patient,” “hospital,” “medicine.”

---

### ⚙️ 🇸🇦 كيف يعمل Skip-Gram؟

### 🇬🇧 How Skip-Gram Works

1. نأخذ كلمة في المنتصف (مثلاً “is”).
2. نحاول توقع الكلمات حولها ضمن نافذة محددة.
3. نكرر العملية لكل كلمة في النص.
4. يتعلم النموذج العلاقات بين الكلمات القريبة بالسياق.

📘 مثال:

> “this is a simple example”
> عند “this”: يتعلم العلاقة مع “is”, “a”.
> عند “is”: يتعلم العلاقة مع “this”, “a”, “simple”.

---

1. Take a target word (e.g., “is”).
2. Predict the words surrounding it within a set window.
3. Repeat for each word in the text.
4. The model learns strong associations between nearby words.

📘 Example:

> “this is a simple example”
> For “this” → predicts “is”, “a”.
> For “is” → predicts “this”, “a”, “simple”.

---

### 💭 🇸🇦 الفرق بين CBOW و Skip-Gram

### 🇬🇧 Difference Between CBOW and Skip-Gram

| 🇸🇦 النموذج  | 🇸🇦 الفكرة             | 🇸🇦 السرعة    | 🇸🇦 الدقة في الكلمات النادرة | 🇬🇧 Model                 | 🇬🇧 Concept | 🇬🇧 Speed                      | 🇬🇧 Rare Word Accuracy |
| ------------- | ----------------------- | -------------- | ----------------------------- | -------------------------- | ------------ | ------------------------------- | ----------------------- |
| **CBOW**      | يتنبأ بالكلمة من السياق | أسرع ⚡         | أقل دقة للكلمات النادرة       | Predicts word from context | Faster ⚡     | Less accurate for rare words    |                         |
| **Skip-Gram** | يتنبأ بالسياق من الكلمة | أبطأ قليلًا 🐢 | أدق للكلمات النادرة 🔍        | Predicts context from word | Slower 🐢    | More accurate for rare words 🔍 |                         |

---

## 🧮 🇸🇦 مشكلة Softmax في Word2Vec

## 🇬🇧 The Softmax Problem

عند وجود مئات الآلاف من الكلمات في القاموس،
تكون دالة **Softmax** بطيئة جدًا لأنها تحسب الاحتمال لكل كلمة ممكنة.

With vocabularies of hundreds of thousands of words,
**Softmax** becomes **very slow and computationally expensive**,
since it must compute probabilities for *every possible word*.

---

## ⚙️ 🇸🇦 حلول لتسريع التدريب

## 🇬🇧 Techniques to Speed Up Training

### 1️⃣ 🇸🇦 الاختيار الهرمي – Hierarchical Softmax

### 1️⃣ 🇬🇧 Hierarchical Softmax

بدل حساب الاحتمال لكل كلمة،
نرتب الكلمات في **شجرة هرمية** (مثل شجرة هوفمان).
الكلمات الشائعة تكون في الأعلى، والنادرة في الأسفل.
الاحتمال النهائي = حاصل ضرب احتمالات المسار للوصول للكلمة.

📘 بهذه الطريقة نحسب احتمالات أقل → أسرع بكثير.

Instead of calculating probabilities for all words,
the vocabulary is arranged in a **hierarchical tree** (like a Huffman tree).
Common words are at the top, rare words at the bottom.
The final probability is the product of probabilities along the path.

📘 This reduces the number of computations, making training much faster.

---

### 2️⃣ 🇸🇦 العينة السلبية – Negative Sampling

### 2️⃣ 🇬🇧 Negative Sampling

الطريقة الأشهر والأسرع ❤️

نختار فقط **عدد صغير من الكلمات الخاطئة** (مثلاً 5–10)
وندرب النموذج على زيادة احتمال الكلمات الصحيحة وتقليل احتمال الخاطئة.

The most common and efficient technique ❤️

Only a **small set of negative (wrong) words** (e.g., 5–10) are sampled,
and the model is trained to increase probabilities for correct words and decrease them for wrong ones.

📘 Example / مثال:

> Input word: “jumps”
> Correct words: {the, fox, brown, over}
> Wrong words: {apple, Tokyo, boat, orange}

النموذج يتعلم أن “jumps” ترتبط بـ “fox” و “over” وليس بـ “apple” أو “Tokyo”.
The model learns that “jumps” relates to “fox” and “over”, but not to “apple” or “Tokyo”.

---

## 🧩 🇸🇦 النتيجة النهائية

## 🇬🇧 Final Outcome

بعد تدريب Word2Vec على كمية ضخمة من النصوص:

* كل كلمة تحصل على **تمثيل رقمي (Vector)** في فضاء المعاني.
* الكلمات المتقاربة في المعنى تكون متقاربة رياضيًا.
* يمكن للنموذج فهم علاقات منطقية مثل:

  > “ملك - رجل + امرأة = ملكة”
  > “مدرسة - طلاب + أطباء ≈ مستشفى”

After training Word2Vec on large text corpora:

* Each word has a **numerical vector** in semantic space.
* Words with similar meanings are close together.
* The model can understand analogies like:

  > “king - man + woman = queen”
  > “school - students + doctors ≈ hospital”

---

## 💡 🇸🇦 خلاصة بسيطة | 🇬🇧 Quick Summary

| 🇸🇦 المفهوم                                     | 🇬🇧 Concept                                                    |
| ------------------------------------------------ | --------------------------------------------------------------- |
| يحوّل الكلمات إلى متجهات رقمية                   | Converts words into numerical vectors                           |
| يتعلم من السياق وليس من التكرار فقط              | Learns from context, not just frequency                         |
| له نوعان: CBOW و Skip-Gram                       | Two main architectures: CBOW and Skip-Gram                      |
| يستخدم Softmax أو Negative Sampling              | Uses Softmax or Negative Sampling                               |
| أساس النماذج الحديثة مثل Word2Vec و GloVe و BERT | Foundation for modern NLP models like Word2Vec, GloVe, and BERT |

---

## ✨ 🇸🇦 ملاحظات مهمة | 🇬🇧 Key Notes

* **CBOW** أسرع في التدريب ومناسب للنصوص الكبيرة.
  **CBOW** trains faster and works well for large datasets.
* **Skip-Gram** أدق في التعامل مع الكلمات النادرة.
  **Skip-Gram** handles rare words with higher accuracy.
* مكتبة **Gensim** في بايثون من أشهر الأدوات لتطبيق Word2Vec.
  **Gensim** is a popular Python library for implementing Word2Vec.
* يمكن تحميل نماذج عربية جاهزة مثل **AraVec**.
  Pre-trained Arabic models like **AraVec** are also available.

---

CBOW:

https://www.kaggle.com/alincijov/continuous-bag-of-words-cbow-numpy-for-beginners

Skip - Gram:

https://www.geeksforgeeks.org/implement-your-own-word2vecskip-gram-model-in-python/

---

# 🧭 🇸🇦 الجزء الخامس: تشابه النصوص | 🇬🇧 Text Similarity 

---

## 1️⃣ 🇸🇦 ما هو تشابه النصوص؟ | 🇬🇧 What is Text Similarity?

**تشابه النصوص (Text Similarity)** يعني:
كيف نقيس *مدى قرب* كلمتين أو جملتين من بعض — إمّا في:

* **الشكل (الحروف)** → مثل: `graffe` و `giraffe`
* **المعنى (السيمانتكس)** → مثل:

  * “الجو جميل اليوم”
  * “اليوم الطقس رائع”

🛠️ تُستخدم هذه الأفكار في أشياء كثيرة، مثل:

* **تصحيح الإملاء (Auto-Correct)**
* **البحث الغامض (Fuzzy Search)**
* **تقييم الترجمة الآلية**
* **كشف الأسئلة/الجمل المتشابهة في الداتا**

---

## 2️⃣ 🇸🇦 تشابه الحروف: مسافة التعديل (Edit Distance) | 🇬🇧 Character-Level Similarity

### 🧱 الفكرة الأساسية

لو عندك كلمة مكتوبة خطأ، مثل:

> `graffe`

والكلمة الصحيحة هي:

> `giraffe`

نريد أن نعرف:
**كم خطوة تعديل نحتاج لنحوّل الأولى إلى الثانية؟**

الخطوات المسموح بها هي ثلاث عمليات:

1. **إضافة حرف (Insertion)**
2. **حذف حرف (Deletion)**
3. **استبدال حرف بحرف آخر (Substitution)**

---

### ✏️ مثال: jraaffe → giraffe

الكلمة الخاطئة: `jraaffe`
الكلمة الصحيحة: `giraffe`

ممكن نعدّلها هكذا:

1. استبدال `j` بـ `g` → `graaffe`
2. إضافة `i` بعد `g` → `giraaffe`
3. حذف حرف `a` الزائد → `giraffe`

إذن **مسافة التعديل (Edit Distance)** = 3
أي نحتاج **3 خطوات** لتحويل الكلمة إلى الصحيحة.

هذا ما يسمّى:

> **Minimum Edit Distance**
> أقل عدد من عمليات: إضافة / حذف / استبدال، لتحويل كلمة إلى أخرى.

---

### 🧮 🇸🇦 كيف تُحسب رياضيًا؟ (ببساطة)

### 🇬🇧 How Is It Computed?

نستخدم خوارزمية تشبه الجداول (Dynamic Programming):

* نضع الكلمة الأولى على الصف العلوي،
* والكلمة الثانية على العمود الجانبي.
* كل خانة تمثل:
  **أقل عدد تعديلات لتحويل جزء من الكلمة الأولى إلى جزء من الثانية.**

نقارن:

* أول حرف مع أول حرف
* أول حرفين مع أول حرفين
* أول 3 حروف مع أول 3 حروف
  … وهكذا، إلى أن ننتهي من أقصر الكلمتين.

في النهاية:
**آخر خانة في الجدول = مسافة التعديل بين الكلمتين.**

ما تحتاج تحفظ الخوارزمية، المهم تفهم الفكرة 👌

---

## 3️⃣ 🇸🇦 وظيفة الـ Edit Distance في الـ Auto-Correct

## 🇬🇧 How It Powers Auto-Correct?

لو كتب المستخدم:

> `graffe`

يقوم النظام بالتالي:

1. يبحث في القاموس عن كلمات قريبة في الحروف (مسافة تعديل صغيرة):

   * `giraffe`
   * `grail`
   * `graft`
   * `graf`
2. يحسب مسافة التعديل بين الكلمة المكتوبة وكل كلمة في القاموس.
3. يختار الكلمات **الأقرب** (أقل Edit Distance).
4. أحيانًا يأخذ في الحسبان:

   * مدى انتشار الكلمة في اللغة
   * سياق الجملة حولها

النتيجة:
يقترح عليك مثلًا:

> Did you mean **“giraffe”**?

---

## 4️⃣ 🇸🇦 تكلفة التعديل (Cost Function) | 🇬🇧 Edit Cost Function

مش كل تعديل لازم يكون له نفس التكلفة.
في بعض التطبيقات:

* حذف حرف = 1
* إضافة حرف = 1
* استبدال حرف = **2** (أصعب شوي)

أو ممكن العكس حسب طبيعة البيانات.

فكرة **Cost Function**:

> نتحكم في وزن كل نوع تعديل،
> عشان نحصل على مسافة أكثر واقعية.

---

## 5️⃣ 🇸🇦 مصفوفة الالتباس بين الحروف | 🇬🇧 Letter Confusion Matrix

بعض الحروف **تتشابه أو تُخلط كثيرًا**:

* في الإنجليزية:

  * `a` ↔ `e`
  * `o` ↔ `u`
* في العربية:

  * `ة` ↔ `ه`
  * `ي` ↔ `ى`

وأحيانًا يكون السبب **قرب الحروف في الكيبورد**:

* مثلًا: `إسلام` ↔ `غسلام`

لذلك بعض الأنظمة تبني **مصفوفة احتمالات**:

> أي حرف غالبًا يُخطأ فيه بأي حرف؟

هذه المصفوفة تُستخدم لتعديل الـ **cost**:
استبدال حرف بحرف "مشابه" يكون **أرخص** من استبداله بحرف بعيد تمامًا.

<p align="center">
  <img src="assets/LCM.png" alt="NLP Tools" width="400">
</p>

---

## 6️⃣ 🇸🇦 تشابه الجمل بالحروف: محدود جدًا

## 🇬🇧 Limitations of Character-Based Similarity

مشكلة كبيرة في تشابه الحروف:

* ممكن جملتين يكون شكلهم متقارب جدًا
  لكن المعنى **مختلف تمامًا**.

مثال:

> “فرنسا دولة غنية بينما ألمانيا يتراجع اقتصادها”
> “ألمانيا دولة غنية بينما فرنسا يتراجع اقتصادها”

نفس الكلمات تقريبًا،
لكن المعنى مقلوب تمامًا.

لذلك:
تشابه الحروف **ممتاز للإملاء**،
لكنه **غير كافٍ لفهم المعنى**.

وهنا نروح للمستوى الأقوى 👇

---

## 7️⃣ 🇸🇦 تشابه المعنى بين الجمل: تمثيل الجمل (Sentence Embeddings)

## 🇬🇧 Semantic Textual Similarity

بدل ما نقيس تشابه **حروف**،
نقيس تشابه **معنى الجملة بالكامل**.

الفكرة:

1. نستخدم نموذج مثل:

   * **Universal Sentence Encoder**
   * أو نماذج أحدث مثل **Sentence-BERT**، أو نماذج تعلّم عميق حديثة.
2. الجملة تدخل كنص:

   > "I love machine learning"
3. تخرج كـ **متجه رقمي طويل** مثلاً:

   * طول 512 قيمة (512-dim vector)
4. جملتين متقاربتين في المعنى →
   المتجهات حقّهم قريبة (Cosine similarity عالية).
   جملتين مختلفتين → المتجهات بعيدة.

---

### 🎛️ 🇸🇦 مصفوفة تشابه الجمل | 🇬🇧 Sentence Similarity Matrix

لو عندك مجموعة جمل، ممكن:

* نحسب تشابه كل جملة مع كل جملة.
* نضع النتائج في **مصفوفة**.
* نرسمها كـ **Heatmap** (ألوان):

  * لون قوي = تشابه عالي
  * لون ضعيف = تشابه قليل

هذا يُستخدم في:

* اكتشاف الجمل المكررة أو المتشابهة.
* فهم أي الجمل تنتمي لنفس الموضوع.
* تصنيف الجمل أو الردود المتقاربة في الـ Chatbots.

<p align="center">
  <img src="assets/Semantic Textual Similarity.png" alt="NLP Tools" width="400">
</p>

<p align="center">
  <img src="assets/USE.png" alt="NLP Tools" width="400">
</p>

---

## 8️⃣ 🇸🇦 2025: ماذا نستخدم اليوم؟

## 🇬🇧 What’s Commonly Used Today?

* على مستوى **الكلمات**:

  * ما زالت **Edit Distance** أساسية في:

    * Auto-correct
    * Fuzzy search
* على مستوى **الجمل والمعنى**:

  * نستخدم **Sentence Embeddings** من نماذج:

    * Sentence-BERT
    * Universal Sentence Encoder
    * نماذج Transformer حديثة (مثل عائلة BERT وGPT وغيرها)

كلها تعتمد على نفس الفكرة:

> تحويل الجملة إلى **متجه** ثم قياس التشابه بينها وبين غيرها باستخدام مقاييس مثل **Cosine Similarity**.

---

# 🧭 🇬🇧 Section 5: Text Similarity

---

## 1️⃣ 🇬🇧 What is Text Similarity?

**Text Similarity** means measuring *how close* two words or sentences are — either in:

* **Form (characters)** → e.g., `graffe` and `giraffe`
* **Meaning (semantics)** → e.g.:

  * “The weather is beautiful today”
  * “Today the weather is lovely”

🛠️ These ideas are widely used in:

* **Spell correction (Auto-Correct)**
* **Fuzzy Search**
* **Machine Translation Evaluation**
* **Detecting similar questions or sentences in data**

---

## 2️⃣ 🇬🇧 Character-Level Similarity: Edit Distance

### 🧱 The Basic Idea

If you have a misspelled word, like:

> `graffe`

and the correct word is:

> `giraffe`

We want to know:
**How many edit steps do we need to transform the first into the second?**

The allowed steps are three operations:

1. **Insertion** — adding a letter
2. **Deletion** — removing a letter
3. **Substitution** — replacing one letter with another

---

### ✏️ Example: jraaffe → giraffe

Misspelled word: `jraaffe`
Correct word: `giraffe`

Possible steps:

1. Replace `j` with `g` → `graaffe`
2. Insert `i` after `g` → `giraaffe`
3. Remove the extra `a` → `giraffe`

So the **Edit Distance = 3**
That means we need **3 steps** to fix the word.

This is called:

> **Minimum Edit Distance**
> The minimum number of Insertions / Deletions / Substitutions needed to transform one word into another.

---

### 🧮 🇬🇧 How Is It Computed? (Simply)

We use an algorithm similar to filling a table — **Dynamic Programming**:

* Place the first word along the top row,
* The second word along the side column.
* Each cell represents:
  **The smallest number of edits needed to turn part of the first word into part of the second.**

We compare:

* First letter vs first letter
* First two letters vs first two letters
* First three vs first three
  … and so on until we reach the shorter word.

Finally:
**The last cell = the total edit distance between the two words.**

You don’t need to memorize the math — just understand the idea 👌

---

## 3️⃣ 🇬🇧 The Role of Edit Distance in Auto-Correct

If a user types:

> `graffe`

The system works like this:

1. Searches in the dictionary for nearby words (small edit distance):

   * `giraffe`
   * `grail`
   * `graft`
   * `graf`
2. Calculates the edit distance between the typed word and each candidate.
3. Picks the **closest** words (with smallest distance).
4. Sometimes considers:

   * How common each word is
   * The sentence context

Then it suggests:

> Did you mean **“giraffe”**?

---

## 4️⃣ 🇬🇧 Edit Cost Function

Not every edit has to cost the same.
In some systems:

* Deleting a character = 1
* Inserting a character = 1
* Substituting = **2** (a bit harder)

Or vice versa — depending on the use case.

The **Cost Function** idea:

> Controls the “weight” of each edit operation
> to get a more realistic distance measurement.

---

## 5️⃣ 🇬🇧 Letter Confusion Matrix

Some letters are **often mixed up**:

* In English:

  * `a` ↔ `e`
  * `o` ↔ `u`
* In Arabic:

  * `ة` ↔ `ه`
  * `ي` ↔ `ى`

Sometimes the cause is **keyboard proximity**:

* For example: `إسلام` ↔ `غسلام`

That’s why some systems build a **probability matrix**:

> Which letters are most often confused with each other?

This matrix is used to adjust the **cost** —
Replacing one letter with a *similar* one is **cheaper** than replacing it with a totally different one.

<p align="center">
  <img src="assets/LCM.png" alt="NLP Tools" width="400">
</p>

---

## 6️⃣ 🇬🇧 Character-Level Sentence Similarity: Very Limited

A major problem with letter-based similarity:

* Two sentences might look *very* similar
  but have **completely different meanings**.

Example:

> “France is a rich country while Germany’s economy is declining.”
> “Germany is a rich country while France’s economy is declining.”

They share almost the same words,
but the meaning is reversed.

So:
Character-level similarity is **great for spelling**,
but **not enough for meaning**.

That’s why we move to a stronger approach 👇

---

## 7️⃣ 🇬🇧 Semantic Sentence Similarity (Sentence Embeddings)

Instead of comparing **letters**,
we compare **sentence meanings**.

The idea:

1. Use a model like:

   * **Universal Sentence Encoder (USE)**
   * or newer ones like **Sentence-BERT**, or other deep-learning Transformers.
2. Input the sentence as text:

   > "I love machine learning"
3. Output: a **numeric vector**, for example:

   * 512-dimensional vector (512 values)
4. If two sentences are semantically close →
   their vectors are close (high cosine similarity).
   If they differ → their vectors are far apart.

---

### 🎛️ 🇬🇧 Sentence Similarity Matrix

If you have multiple sentences, you can:

* Compute similarity for every pair of sentences.
* Store results in a **matrix**.
* Visualize it as a **heatmap** (color-coded):

  * Strong color = high similarity
  * Faint color = low similarity

This is useful for:

* Detecting duplicate or similar sentences.
* Understanding which sentences belong to the same topic.
* Grouping or classifying chatbot responses.

<p align="center">
  <img src="assets/Semantic Textual Similarity.png" alt="NLP Tools" width="400">
</p>

<p align="center">
  <img src="assets/USE.png" alt="NLP Tools" width="400">
</p>

---

## 8️⃣ 🇬🇧 What’s Commonly Used Today (2025)

* On the **word level**:

  * **Edit Distance** remains essential for:

    * Auto-correct
    * Fuzzy Search
* On the **sentence/meaning level**:

  * We use **Sentence Embeddings** from models like:

    * **Sentence-BERT**
    * **Universal Sentence Encoder (USE)**
    * Modern Transformer models (e.g., **BERT**, **GPT**, etc.)

They all follow the same idea:

> Convert a sentence into a **vector**,
> then measure similarity using metrics such as **Cosine Similarity**.

---

## 🪄 🇬🇧 Quick Summary

| 🇬🇧 Concept                       | Description                                |
| ---------------------------------- | ------------------------------------------ |
| **Edit Distance**                  | Measures minimal edits between words       |
| **Cost Function**                  | Assigns weights to edit operations         |
| **Letter Confusion Matrix**        | Models how letters are often confused      |
| **Character Similarity**           | Good for spelling, not for meaning         |
| **Sentence Embeddings (512 dims)** | Capture **semantic** similarity            |
| **Heatmap / Similarity Matrix**    | Visualizes sentence-to-sentence similarity |

---



# 🧭 🇸🇦 الجزء السادس والأخير: التشابه التوزيعي | 🇬🇧 Distributional Similarity

---

## 1️⃣ 🇸🇦 ما هو التشابه التوزيعي؟ | 🇬🇧 What Is Distributional Similarity?

### 🇸🇦 بالعربي:

التشابه التوزيعي فكرة ذكية تُستخدم لفهم معنى الكلمات **بدون قاموس**.
لأن القواميس ما تغطي كل اللغات أو الكلمات الجديدة.
فالفكرة تقول:

> "الكلمة تُعرف من سياقها — أي من الكلمات التي تظهر بجانبها."

يعني لو ظهرت كلمة غريبة كثيرًا بجانب كلمات نعرفها، نقدر نخمن معناها من البيئة حولها.

### 🇬🇧 In English:

Distributional Similarity helps us understand **word meanings without a dictionary**.
Since dictionaries can’t include every new word, we rely on context:

> “You shall know a word by the company it keeps.”

If an unknown word appears near known ones, we can infer its meaning from the surrounding context.

---

## 2️⃣ 🇸🇦 مثال بسيط | 🇬🇧 A Simple Example

### 🇸🇦 بالعربي:

لو عندنا الجمل التالية:

1. ذبح أبي **قنعرا**
2. لم يحب أخي لحم **القنعر** البارحة
3. هرول **القنعر** هربًا من الليث
4. إنه سمين كـ **القنعر**

حتى لو ما نعرف كلمة "القنعر"، نلاحظ أنها تظهر مع كلمات مثل:
"ذبح"، "لحم"، "سمين"، "هرب"
فنستنتج أنها تشير إلى **حيوان سمين**.

وهذا بالضبط هو جوهر **التشابه التوزيعي**.

### 🇬🇧 In English:

Suppose we have these sentences:

1. My father slaughtered a **qanʿar**.
2. My brother disliked **qanʿar** meat yesterday.
3. The **qanʿar** ran away from the lion.
4. It’s as fat as a **qanʿar**.

Even without knowing “qanʿar,”
we can infer it’s likely **an animal**, probably a **fat one**,
since it appears with words like *meat*, *lion*, and *fat*.

That’s **distributional similarity** — inferring meaning from context.

---

## 3️⃣ 🇸🇦 الفكرة الرياضية | 🇬🇧 The Mathematical Intuition

### 🇸🇦 بالعربي:

نخزن الكلمات في **مصفوفة كبيرة** تمثل كم مرة تظهر كلمة بجانب كلمة أخرى.

<p align="center">
  <img src="assets/dist-similarity.png" alt="PPMI Example" width="400">
</p>

<p align="center">
  <img src="assets/dist-similarity-2.png" alt="Distributional Similarity" width="400">
</p>

<p align="center">
  <img src="assets/dist-similarity-3.png" alt="Distributional Similarity" width="400">
</p>


### 🇬🇧 In English:

We represent words in a **co-occurrence matrix** —
how often each word appears near others.

<p align="center">
  <img src="assets/dist-similarity.png" alt="PPMI Example" width="400">
</p>

<p align="center">
  <img src="assets/dist-similarity-2.png" alt="Distributional Similarity" width="400">
</p>

<p align="center">
  <img src="assets/dist-similarity-3.png" alt="Distributional Similarity" width="400">
</p>

---

## 4️⃣ 🇸🇦 النطاق (Window) | 🇬🇧 Context Window

### 🇸🇦 بالعربي:

ما نحتاج نحلل النص كله،
بل نكتفي بعدد كلمات قبل وبعد كل كلمة — مثلًا **20 كلمة** —
لأن هذا يكفي لتحديد معناها التقريبي.

### 🇬🇧 In English:

We don’t need to scan the whole text.
We use a **context window** — say, 20 words around each target word —
to capture the nearby meaning effectively.

---

## 5️⃣ 🇸🇦 من الجمل إلى الأرقام | 🇬🇧 From Sentences to Numbers

### 🇸🇦 بالعربي:

نحوّل النصوص إلى مصفوفات رقمية تمثل كم مرة تظهر الكلمات معًا.
مثلًا:
`pineapple` و `apricot` تظهران في أماكن متقاربة،
بينما `digital` و `information` كذلك متقاربتان.

فنقول:

> الكلمات المتشابهة في التوزيع → متشابهة في المعنى.

### 🇬🇧 In English:

We convert text into numeric matrices showing co-occurrence frequencies.
If `pineapple` and `apricot` appear in similar contexts,
and `digital` with `information`,
then they share similar meanings.

> Words that appear in similar contexts have similar meanings.

---


## 6️⃣ 🇸🇦 خوارزمية PPMI — متى نستخدمها وليش

## 6️⃣ 🇬🇧 PPMI Algorithm — When and Why We Use It

---

### 🇸🇦 أولًا: ليش نحتاجها؟

لما نكوّن **مصفوفة التشارك (co-occurrence matrix)** بين الكلمات والسياقات،
بيكون عندنا أرقام بسيطة مثل:

> كلمة "book" ظهرت مع "read" = 5 مرات، ومع "pen" = 2 مرات.

لكن مجرد **عدد الظهور الخام (count)** ما يكفي 😕
لأنه ممكن كلمة "the" تظهر مع كل شيء تقريبًا — فهل هذا يعني إنها مهمة؟ لا.

إحنا نحتاج نعرف:

> هل ظهور الكلمة بجانب كلمة معينة “مميز” فعلًا؟ ولا مجرد صدفة؟

وهنا يجي دور **PPMI** 💡

---

### 🇬🇧 Why We Need PPMI

Raw word counts (like how often two words co-occur) aren’t meaningful on their own.
Common words like *the*, *and*, *a* appear everywhere — but they don’t tell us real relationships.

We need to know:

> Is this co-occurrence **statistically significant** or just random?

That’s what **PPMI (Positive Pointwise Mutual Information)** measures.

---

### 🇸🇦 ثانيًا: متى نستخدمها

نستخدم PPMI في أي وقت نبغى **نقيس العلاقة الحقيقية بين الكلمة والسياق**
بدون ما نتأثر بالكلمات الشائعة جدًا.

🧠 أمثلة لاستخدامها:

1. بناء **تمثيلات الكلمات (word embeddings)** من نصوص ضخمة.
2. تحليل **تشابه الكلمات** (مثل Word2Vec لكن بطريقة إحصائية).
3. إنشاء **خرائط دلالية** أو **مصفوفات توزيع** لفهم العلاقات بين المفاهيم.

---

### 🇬🇧 When We Use It

We use **PPMI** whenever we want to measure how strongly a word is associated with its context,
without being biased by very frequent words.

✅ Use cases:

1. Building **word embeddings** from co-occurrence matrices.
2. Measuring **word similarity** statistically.
3. Creating **semantic maps** or **distributional representations**.

---

### 🇸🇦 ثالثًا: وش تسوي بالضبط

PPMI تعيد “وزن” العلاقات داخل المصفوفة بحيث:

* العلاقة القوية بين كلمتين نادرة لكنها مرتبطة جدًا → **قيمة عالية**
* العلاقة الشائعة بين كلمتين غير مميزتين → **قيمة منخفضة أو صفر**

وهذا يخلي المصفوفة تعبّر عن **المعنى** بدل التكرار فقط.

---

### 🇬🇧 What It Actually Does

PPMI **re-weights** the raw co-occurrence matrix:

* Rare but meaningful associations → **High score**
* Common but uninformative ones → **Low or zero**

This helps capture **true semantic relationships**, not just word frequency.

---

### 🇸🇦 مثال سريع

كلمة "king" تظهر كثير مع "queen" أكثر من "banana"،
حتى لو "banana" ظهرت كثير بشكل عام.
PPMI يفهم إن العلاقة "king–queen" أكثر **دلالية** من مجرد التكرار.

---

### 🇬🇧 Quick Example

“king” co-occurs with “queen” more meaningfully than with “banana,”
even if “banana” appears more often overall.
PPMI captures that — it focuses on **semantic significance**, not raw counts.

---

## 7️⃣ 🇸🇦 مشكلة الـ Bias وحل لابلاس | 🇬🇧 Bias Problem and Laplace Smoothing

### 🇸🇦 بالعربي:

أحيانًا تظهر القيم غير منطقية —
مثلاً يعطي تشابه عالي بين `pineapple` و `apricot`
لكن منخفض بين `digital` و `information`.

السبب؟
وجود **تحيّز (bias)** في الحسابات عندما تكون بعض القيم صغيرة جدًا.

الحل؟
نضيف رقم صغير (مثل **2**) لكل القيم قبل الحساب.
هذا يُسمّى **Laplace Smoothing**.

النتيجة:
تصبح الأرقام أكثر منطقية ومتوازنة.

### 🇬🇧 In English:

Sometimes results look illogical —
e.g., `pineapple` and `apricot` show higher similarity than `digital` and `information`.

The reason: statistical **bias** from low-frequency counts.

The fix:
Add a small constant (like **2**) to all counts before computing.
This is called **Laplace Smoothing**.

This adjustment produces more realistic similarity scores.

---

## 8️⃣ 🇸🇦 الخلاصة | 🇬🇧 Summary

| 🇸🇦 المفهوم                   | 🇬🇧 Concept                             |
| ------------------------------ | ---------------------------------------- |
| التشابه التوزيعي               | Distributional Similarity                |
| الكلمة تُفهم من سياقها         | A word is understood by its context      |
| مصفوفة التشارك (Co-occurrence) | Measures how often words appear together |
| PPMI                           | Measures strength of association         |
| Laplace Smoothing              | Fixes bias in frequency counts           |


---
